In [1]:
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', None)

## Documentation

For a comprehensive list describing all the features/statistics in this notebook; refer to the following documentations links:

- <a href = 'https://library.fangraphs.com/offense/offensive-statistics-list/'> FanGraphs Batting </a> 

- <a href = 'https://library.fangraphs.com/pitching/complete-list-pitching/'> FanGraphs Pitching </a> 

## Batting Data

The `batting_stats` function returns basic and advanced season-level batting data from FanGraphs. 

While this query should work for any historical season, some of the more modern stats (contact %, zone %, and many others) will not be available before certain dates. 

*A note about the statistics being queried: most of the advanced level statistics will not be available before 2014 because <a href = 'https://www.mlb.com/glossary/statcast'> StatCast </a> was not introduced yet.*

In [2]:
from pybaseball import batting_stats

In [3]:
# retrieve data on only players who have 100+ plate appearances between 2000 and 2021
# batting_2021 = batting_stats(2000, 2021, qual=100)

In [4]:
# load in csv
batting_2021 = pd.read_csv('Data/batting_2021', index_col = 0)

In [5]:
batting_2021.head()

IDfg  Season          Name Team  Age    G   AB   PA    H  1B  2B  3B  HR  \
0    1109    2002   Barry Bonds  SFG   37  143  403  612  149  70  31   2  46   
2    1109    2001   Barry Bonds  SFG   36  153  476  664  156  49  32   2  73   
1    1109    2004   Barry Bonds  SFG   39  147  373  617  135  60  27   3  45   
38  13611    2018  Mookie Betts  BOS   25  136  520  614  180  96  47   5  32   
3    1109    2003   Barry Bonds  SFG   38  130  390  550  133  65  22   1  45   

      R  RBI   BB  IBB  SO  HBP  SF  SH  GDP  SB  CS    AVG     GB     FB  \
0   117  110  198   68  47    9   2   0    4   9   2  0.370  109.0  169.0   
2   129  137  177   35  93    9   2   0    5  13   3  0.328    NaN    NaN   
1   129  101  232  120  41    9   3   0    5   6   1  0.362  117.0  156.0   
38  129   80   81    8  91    8   5   0    5  30   6  0.346  147.0  195.0   
3   111   90  148   61  58   10   2   0    7   7   0  0.341  101.0  152.0   

      LD  IFFB  Pitches   Balls  Strikes   IFH   BU  BUH    BB%     K%  BB/K  \
0   80.0  16.0   2398.0  1336.0   1062.0   5.0  0.0  0.0  0.324  0.077  4.21   
2    NaN   NaN      NaN     NaN      NaN   NaN  NaN  NaN  0.267  0.140  1.90   
1   62.0  22.0   2425.0  1444.0    981.0   8.0  0.0  0.0  0.376  0.066  5.66   
38  92.0  18.0   2582.0  1095.0   1487.0  14.0  0.0  0.0  0.132  0.148  0.89   
3   81.0  21.0   2175.0  1141.0   1034.0   7.0  0.0  0.0  0.269  0.105  2.55   

      OBP    SLG    OPS    ISO  BABIP   GB/FB    LD%    GB%    FB%  IFFB%  \
0   0.582  0.799  1.381  0.429  0.330  0.0064  0.223  0.304  0.472  0.095   
2   0.515  0.863  1.379  0.536  0.266     NaN    NaN    NaN    NaN    NaN   
1   0.609  0.812  1.422  0.450  0.310  0.0075  0.185  0.349  0.466  0.141   
38  0.438  0.640  1.078  0.294  0.368  0.0075  0.212  0.339  0.449  0.092   
3   0.529  0.749  1.278  0.408  0.304  0.0066  0.243  0.302  0.455  0.138   

    HR/FB   IFH%  BUH%   wOBA   wRAA  wRC    Bat   Fld   Rep  Pos    RAR  \
0   0.272  0.046   0.0  0.544  110.0  183  110.1   4.3  18.6 -6.3  126.8   
2     NaN    NaN   NaN  0.537  117.7  200  116.8  -5.0  20.7 -7.0  127.8   
1   0.288  0.068   0.0  0.537  108.0  185  106.0   2.2  19.1 -6.6  122.1   
38  0.164  0.095   0.0  0.449   67.2  139   62.4  18.5  18.4 -5.3  103.3   
3   0.296  0.069   0.0  0.503   80.5  148   79.3  11.6  17.0 -6.0  103.6   

     WAR    Dol  Spd  wRC+    WPA  -WPA   +WPA    RE24    REW   pLI  phLI  PH  \
0   12.7  $49.5  3.9   244  10.51 -5.90  16.40  118.51  12.63  0.98  3.23   4   
2   12.5    NaN  4.1   235  11.63 -7.46  19.10  118.93  12.11  0.97  1.34   5   
1   11.9  $49.9  4.0   233  13.06 -5.23  18.29  128.94  12.86  1.04  2.18   8   
38  10.6  $85.1  7.0   185   5.77 -8.16  13.93   62.93   6.11  0.92  3.62   4   
3   10.2  $43.9  3.9   212   7.99 -5.94  13.93   84.65   8.61  0.99  2.36   2   

    WPA/LI  Clutch  FB% (Pitch)   FBv    SL%   SLv    CT%   CTv    CB%   CBv  \
0    11.82   -1.11        0.678  89.0  0.108  79.9    NaN   NaN  0.081  74.2   
2    13.04   -1.04          NaN   NaN    NaN   NaN    NaN   NaN    NaN   NaN   
1    11.35    1.15        0.607  90.1  0.136  82.7  0.005  88.3  0.094  75.2   
38    6.44   -0.15        0.547  93.2  0.199  84.5  0.050  88.7  0.073  79.1   
3     8.38   -0.34        0.629  89.5  0.128  80.3    NaN   NaN  0.088  74.9   

      CH%   CHv    SF%   SFv    KN%   KNv    XX%    PO%   wFB   wSL  wCT  \
0   0.107  76.5  0.026  80.5    NaN   NaN  0.081    NaN  67.8  12.5  NaN   
2     NaN   NaN    NaN   NaN    NaN   NaN    NaN    NaN   NaN   NaN  NaN   
1   0.129  81.9  0.022  80.6  0.008  74.2  0.184  0.059  52.6   8.3  1.7   
38  0.114  84.4  0.017  85.4    NaN   NaN  0.009    NaN  40.1  11.2  4.2   
3   0.135  80.5  0.019  82.6  0.001   NaN  0.109    NaN  52.6   3.5  NaN   

     wCB   wCH  wSF  wKN  wFB/C  wSL/C  wCT/C  wCB/C  wCH/C  wSF/C  wKN/C  \
0    5.1  19.2  0.8  NaN   4.54   5.27    NaN   2.84   8.13   1.37    NaN   
2    NaN   NaN  NaN  NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  

In [6]:
batting_2021.describe()

IDfg        Season           Age             G            AB  \
count  1.200600e+04  12006.000000  12006.000000  12006.000000  12006.000000   
mean   1.968213e+04   2010.426287     28.612360     89.841163    287.096535   
std    1.194781e+05      6.292497      4.150018     45.450986    186.734319   
min    1.000000e+00   2000.000000     19.000000     12.000000     36.000000   
25%    1.036250e+03   2005.000000     25.000000     45.000000    112.000000   
50%    3.311000e+03   2010.000000     28.000000     90.000000    252.000000   
75%    9.063000e+03   2016.000000     31.000000    133.000000    456.000000   
max    1.014462e+06   2021.000000     48.000000    163.000000    716.000000   

                 PA             H            1B            2B            3B  \
count  12006.000000  12006.000000  12006.000000  12006.000000  12006.000000   
mean     321.890222     75.034566     49.217475     15.011827      1.537731   
std      209.826797     54.883957     36.203340     11.825252      2.104999   
min       50.000000      0.000000      0.000000      0.000000      0.000000   
25%      124.250000     26.000000     17.000000      5.000000      0.000000   
50%      281.500000     63.000000     41.000000     12.000000      1.000000   
75%      510.000000    120.000000     77.000000     24.000000      2.000000   
max      778.000000    262.000000    225.000000     59.000000     23.000000   

                 HR             R           RBI            BB           IBB  \
count  12006.000000  12006.000000  12006.000000  12006.000000  12006.000000   
mean       9.267533     38.859487     37.125354     27.449858      1.996335   
std        9.829529     30.438054     30.640863     23.689191      3.634348   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        2.000000     13.000000     11.000000      9.000000      0.000000   
50%        6.000000     31.000000     29.000000     21.000000      1.000000   
75%       14.000000     61.000000     57.000000     40.000000      3.000000   
max       73.000000    152.000000    160.000000    232.000000    120.000000   

                 SO           HBP            SF            SH           GDP  \
count  12006.000000  12006.000000  12006.000000  12006.000000  12006.000000   
mean      60.077794      3.039064      2.275612      1.979927      6.479594   
std       39.375700      3.574619      2.344524      2.946817      5.564244   
min        2.000000      0.000000      0.000000      0.000000      0.000000   
25%       27.000000      0.000000      0.000000      0.000000      2.000000   
50%       52.000000      2.000000      2.000000      1.000000      5.000000   
75%       85.000000      4.000000      3.000000      3.000000     10.000000   
max      223.000000     31.000000     16.000000     24.000000     32.000000   

                 SB            CS           AVG            GB            FB  \
count  12006.000000  12006.000000  12006.000000  10901.000000  10901.000000   
mean       4.790105      1.881309      0.242474     98.832492     80.735162   
std        7.994766      2.619757      0.052402     69.581627     59.382214   
min        0.000000      0.000000      0.000000      2.000000      0.000000   
25%        0.000000      0.000000      0.218000     37.000000     28.000000   
50%        2.000000      1.000000      0.251000     84.000000     68.000000   
75%        6.000000      3.000000      0.277000    151.000000    126.000000   
max       78.000000     24.000000      0.397000    401.000000    285.000000   

                 LD          IFFB       Pitches         Balls       Strikes  \
count  10901.000000  10901.000000  10901.000000  10901.000000  10901.000000   
mean      46.287130      8.217503   1228.572791    454.508302    774.064489   
std       33.773407      7.468154    809.704250    316.561484    498.733437   
min        0.000000      0.000000    136.000000     22.000000     94.000000   
25%       16.000000      2.000000    476.000000    169.00000

## bWAR Batting

Get Baseball Reference's WAR stats from its war_daily_bat table, along with some other data that's not included in the batting_stats_bref function. 

For the purposes of this project, we will care about salary. We can later merge this dataframe with the `batting_2022` dataframe.

In [7]:
from pybaseball import bwar_bat

# get war stats from baseball reference 
bwar_bat = bwar_bat()

# remove excess stats; just need salary
cols_to_remove = ['runs_above_avg','runs_above_avg_off', 'runs_above_avg_def', 
                  'WAR_rep', 'WAA', 'WAR']

bwar_bat = bwar_bat.drop(columns = cols_to_remove)
bwar_bat.head()

name_common    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID  \
0  David Aardsma  430911.0  aardsda01     2004     SFG         1    NL   
1  David Aardsma  430911.0  aardsda01     2006     CHC         1    NL   
2  David Aardsma  430911.0  aardsda01     2007     CHW         1    AL   
3  David Aardsma  430911.0  aardsda01     2008     BOS         1    AL   
4  David Aardsma  430911.0  aardsda01     2009     SEA         1    AL   

  pitcher   G   PA    salary  
0       Y  11  0.0  300000.0  
1       Y  43  3.0       NaN  
2       Y   2  0.0  387500.0  
3       Y   5  1.0  403250.0  
4       Y   3  0.0  419000.0

In [8]:
# save as csv
path = "Data/"

bwar_bat.to_csv(path + 'bwar_bat')

In [9]:
# load in csv
bwar_bat = pd.read_csv('Data/bwar_bat', index_col = 0)

In [10]:
# filter bwar_bat to 2000 - 2021 and minimum 100 PA
bwar_bat_2021 = bwar_bat[(bwar_bat['year_ID'].between(2000,2021)) & (bwar_bat['PA'] >= 100)]
bwar_bat_2021.sort_values(by = 'year_ID')

name_common    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID  \
109         Jeff Abbott  110010.0  abbotje01     2000     CHW         1    AL   
39743     Alex Gonzalez  136460.0  gonzaal02     2000     FLA         1    NL   
39888     Juan Gonzalez  114932.0  gonzaju03     2000     DET         1    AL   
39907     Luis Gonzalez  114935.0  gonzalu01     2000     ARI         1    NL   
40034     Wiki Gonzalez  150435.0  gonzawi01     2000     SDP         1    NL   
...                 ...       ...        ...      ...     ...       ...   ...   
80075     Shohei Ohtani  660271.0  ohtansh01     2021     LAA         1    AL   
80195   Edward Olivares  658668.0  olivaed02     2021     KCR         1    AL   
80462        Matt Olson  621566.0  olsonma02     2021     OAK         1    AL   
77476       Josh Naylor  647304.0  naylojo01     2021     CLE         1    AL   
119788      Mike Zunino  572287.0  zuninmi01     2021     TBR         1    AL   

       pitcher    G     PA     salary  
109          N   80  242.0   255000.0  
39743        N  109  407.0   295000.0  
39888        N  115  496.0  7500000.0  
39907        N  162  722.0  3333333.0  
40034        N   95  319.0   208000.0  
...        ...  ...    ...        ...  
80075        N  155  639.0  3000000.0  
80195        N   39  111.0        NaN  
80462        N  156  673.0  5000000.0  
77476        N   69  250.0        NaN  
119788       N  109  375.0  2000000.0  

[9925 rows x 11 columns]

In [11]:
bwar_bat_2021.describe()

mlb_ID      year_ID     stint_ID            G           PA  \
count    9925.000000  9925.000000  9925.000000  9925.000000  9925.000000   
mean   375076.583476  2010.392343     1.059950    99.661562   368.756272   
std    181475.811853     6.303315     0.246157    39.791636   188.212288   
min    110010.000000  2000.000000     1.000000    22.000000   100.000000   
25%    150029.000000  2005.000000     1.000000    63.000000   197.000000   
50%    430906.000000  2010.000000     1.000000   101.000000   343.000000   
75%    502517.000000  2016.000000     1.000000   137.000000   536.000000   
max    683734.000000  2021.000000     4.000000   163.000000   778.000000   

             salary  
count  7.789000e+03  
mean   4.198015e+06  
std    5.359650e+06  
min    1.000000e+05  
25%    5.170000e+05  
50%    1.880000e+06  
75%    6.000000e+06  
max    3.766667e+07

In [12]:
# how many missing values in batting salary column?
print("Missing bWAR Batting Salary Data:")
bwar_bat_2021['salary'].isna().value_counts(normalize = True)

Missing bWAR Batting Salary Data:


False    0.784786
True     0.215214
Name: salary, dtype: float64

## Pitching Data

The `pitching_stats` function returns basic and advanced season-level pitching data from FanGraphs.

While this query should work for any historical season, some of the more modern stats (contact %, zone %, and many others) will not be available before certain dates.

In [13]:
from pybaseball import pitching_stats

In [14]:
# retrieve data on only players who have pitched 30+ innings since 2000
pitching_2021 = pitching_stats(2000, 2021, qual=30)

In [15]:
# load in csv
pitching_2021 = pd.read_csv('Data/pitching_2021', index_col = 0)

In [16]:
pitching_2021.head()

IDfg  Season            Name Team  Age   W   L   WAR   ERA   G  GS  CG  \
910     60    2001   Randy Johnson  ARI   37  21   6  10.4  2.49  35  34   3   
1152    60    2000   Randy Johnson  ARI   36  19   7   9.6  2.64  35  35   8   
1062    60    2004   Randy Johnson  ARI   40  16  14   9.6  2.60  35  35   4   
209    200    2000  Pedro Martinez  BOS   28  18   6   9.4  1.74  29  29   7   
2393    73    2002  Curt Schilling  ARI   35  23   7   9.3  3.23  36  35   5   

      ShO  SV   BS     IP   TBF    H   R  ER  HR  BB  IBB  HBP  WP  BK   SO  \
910     2   0  NaN  249.2   994  181  74  69  19  71    2   18   8   1  372   
1152    3   0  NaN  248.2  1001  202  89  73  23  76    1    6   5   2  347   
1062    2   0  0.0  245.2   964  177  88  71  18  44    1   10   3   1  290   
209     4   0  NaN  217.0   817  128  44  42  17  32    0   14   1   0  284   
2393    1   0  0.0  259.1  1017  218  95  93  29  33    1    3   6   0  316   

         GB     FB     LD  IFFB   Balls  Strikes  Pitches     RS   IFH    BU  \
910     NaN    NaN    NaN   NaN     NaN      NaN      NaN    NaN   NaN   NaN   
1152    NaN    NaN    NaN   NaN     NaN      NaN      NaN    NaN   NaN   NaN   
1062  261.0  224.0  120.0  24.0  1125.0   2508.0   3633.0  109.0  21.0  15.0   
209     NaN    NaN    NaN   NaN     NaN      NaN      NaN    NaN   NaN   NaN   
2393  272.0  235.0  143.0  26.0  1092.0   2630.0   3722.0  171.0  16.0  15.0   

      BUH    K/9  BB/9  K/BB   H/9  HR/9    AVG  WHIP  BABIP   LOB%   FIP  \
910   NaN  13.41  2.56  5.24  6.52  0.68  0.200  1.01  0.315  0.805  2.13   
1152  NaN  12.56  2.75  4.57  7.31  0.83  0.220  1.12  0.326  0.774  2.53   
1062  4.0  10.62  1.61  6.59  6.48  0.66  0.195  0.90  0.264  0.695  2.30   
209   NaN  11.78  1.33  8.88  5.31  0.71  0.166  0.74  0.236  0.866  2.17   
2393  3.0  10.97  1.15  9.58  7.57  1.01  0.222  0.97  0.297  0.745  2.40   

      GB/FB    LD%    GB%    FB%  IFFB%  HR/FB   IFH%   BUH%  Starting  \
910     NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN      86.4   
1152    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN      90.1   
1062   1.17  0.198  0.431  0.370  0.107  0.080  0.080  0.267      86.1   
209     NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN      87.1   
2393   1.16  0.220  0.418  0.362  0.111  0.123  0.059  0.200      81.9   

      Start-IP  Relieving  Relief-IP   RAR Dollars  tERA  xFIP   WPA   -WPA  \
910      242.2        4.5        7.0  90.9     NaN   NaN   NaN  6.23 -14.23   
1152     248.2        NaN        NaN  90.1     NaN   NaN   NaN  5.38 -15.87   
1062     245.2        NaN        NaN  86.1   $40.3  2.72  2.64  4.00 -14.28   
209      217.0        NaN        NaN  87.1     NaN   NaN   NaN  7.61  -9.24   
2393     258.1       -0.5        1.0  81.3   $36.3  3.10  2.20  4.63 -13.43   

       +WPA   RE24   REW   pLI  inLI  gmLI  exLI  Pulls  WPA/LI  Clutch  \
910   20.46  64.42  6.25  1.04  0.98  0.88  1.07     31    5.84    0.17   
1152  21.24  50.15  4.82  1.08  0.99  0.87  1.15     27    4.41    0.57   
1062  18.28  48.07  4.59  0.98  0.99  0.87  1.38     31    6.25   -2.16   
209   16.86  80.32  7.51  0.97  0.97  0.87  1.05     22    8.09   -0.25   
2393  18.06  43.62  4.27  0.90  0.93  0.84  1.23     31    6.01   -0.88   

      FB% 2   FBv    SL%   SLv  CT%  CTv    CB%   CBv    CH%   CHv    SF%  \
910     NaN   NaN    NaN   NaN  NaN  NaN    NaN   NaN    NaN   NaN    NaN   
1152    NaN   NaN    NaN   NaN  NaN  NaN    NaN   NaN    NaN   NaN    NaN   
1062  0.453  93.7  0.436  84.4  NaN  NaN    NaN   NaN  0.067  85.1  0.045   
209     NaN   NaN    NaN   NaN  NaN  NaN    NaN   NaN    NaN   NaN    NaN   
2393  0.552  92.7  0.126  85.6  NaN  NaN  0.055  75.5  0.042  83.4  0.224   

       SFv  KN%  KNv    XX%    PO%   wFB   wSL  wCT  wCB  wCH   wSF  wKN  \
910    NaN  NaN  NaN    NaN    NaN   NaN   NaN  NaN  NaN  NaN   NaN  NaN   
1152   NaN  NaN  NaN    NaN    NaN   NaN   NaN  NaN  NaN  NaN   NaN  NaN   
1062  85.7  NaN  NaN  0.098  0.001  11.0  33

In [17]:
pitching_2021.describe()

IDfg        Season           Age             W             L  \
count  1.051900e+04  10519.000000  10519.000000  10519.000000  10519.000000   
mean   1.975712e+04   2010.736287     28.348988      4.818994      4.739519   
std    1.173236e+05      6.307353      4.166458      4.397048      3.761656   
min    3.000000e+00   2000.000000     19.000000      0.000000      0.000000   
25%    1.138000e+03   2005.000000     25.000000      2.000000      2.000000   
50%    4.138000e+03   2011.000000     28.000000      3.000000      4.000000   
75%    9.810000e+03   2016.000000     31.000000      7.000000      7.000000   
max    1.014446e+06   2021.000000     49.000000     24.000000     21.000000   

                WAR           ERA             G            GS            CG  \
count  10519.000000  10519.000000  10519.000000  10519.000000  10519.000000   
mean       0.897604      4.325571     36.068257      9.703964      0.271604   
std        1.359167      1.492954     20.009926     12.012686      0.830297   
min       -1.600000      0.000000      4.000000      0.000000      0.000000   
25%        0.000000      3.320000     21.000000      0.000000      0.000000   
50%        0.500000      4.180000     32.000000      3.000000      0.000000   
75%        1.400000      5.155000     52.000000     19.000000      0.000000   
max       10.400000     13.140000     94.000000     36.000000     11.000000   

                ShO            SV           BS            IP           TBF  \
count  10519.000000  10519.000000  9638.000000  10519.000000  10519.000000   
mean       0.112558      2.438350     1.142457     84.111237    360.855405   
std        0.412157      7.643958     1.826539     58.102023    243.855154   
min        0.000000      0.000000     0.000000     20.000000     75.000000   
25%        0.000000      0.000000     0.000000     40.200000    179.000000   
50%        0.000000      0.000000     0.000000     64.000000    271.000000   
75%        0.000000      1.000000     2.000000    115.000000    499.000000   
max        6.000000     62.000000    12.000000    266.000000   1071.000000   

                  H             R            ER            HR            BB  \
count  10519.000000  10519.000000  10519.000000  10519.000000  10519.000000   
mean      82.872802     42.376652     39.048294     10.047818     29.913395   
std       58.752005     29.629011     27.437827      7.773112     19.314587   
min        7.000000      0.000000      0.000000      0.000000      1.000000   
25%       39.000000     20.000000     19.000000      4.000000     15.000000   
50%       59.000000     31.000000     28.000000      7.000000     24.000000   
75%      118.000000     62.000000     57.000000     14.000000     40.000000   
max      271.000000    152.000000    145.000000     48.000000    125.000000   

                IBB           HBP            WP            BK            SO  \
count  10519.000000  10519.000000  10519.000000  10519.000000  10519.000000   
mean       2.151630      3.312007      3.096112      0.291473     68.996958   
std        2.086209      3.005256      2.840299      0.619634     49.658313   
min        0.000000      0.000000      0.000000      0.000000      4.000000   
25%        1.000000      1.000000      1.000000      0.000000     32.000000   
50%        2.000000      3.000000      2.000000      0.000000     54.000000   
75%        3.000000      5.000000      4.000000      0.000000     90.000000   
max       17.000000     24.000000     25.000000      6.000000    372.000000   

                GB           FB           LD         IFFB        Balls  \
count  9638.000000  9638.000000  9638.000000  9638.000000  9638.000000   
mean    109.963893    88.565678    50.812202     9.064640   499.914920   
std      82.919192    64.477807    37.016805     7.723648   329.168027   
min       6.000000     7.000000     1.000000     0.000000    86.000000   
25%      49.000000    41.000000    24.000000     4.000000   255.000000   
50%      80.000000  

## bWAR Pitching 

Get Baseball Reference's WAR stats from its war_daily_pitch table, along with some other data that's not included in the pitching_stats_bref function. 

For the purposes of this project, we will care about salary. We can later merge this dataframe with the `pitching_2022` dataframe.

In [18]:
from pybaseball import bwar_pitch

# get war stats from baseball reference 
bwar_pitch = bwar_pitch(return_all = True)

# remove excess stats; just need salary
cols_to_keep = ['name_common', 'age', 'player_ID', 'year_ID', 'team_ID',
                'stint_ID', 'lg_ID', 'salary']

bwar_pitch = bwar_pitch[cols_to_keep]
bwar_pitch.head()

name_common   age  player_ID  year_ID team_ID  stint_ID lg_ID    salary
0  David Aardsma  22.0  aardsda01     2004     SFG         1    NL  300000.0
1  David Aardsma  24.0  aardsda01     2006     CHC         1    NL       NaN
2  David Aardsma  25.0  aardsda01     2007     CHW         1    AL  387500.0
3  David Aardsma  26.0  aardsda01     2008     BOS         1    AL  403250.0
4  David Aardsma  27.0  aardsda01     2009     SEA         1    AL  419000.0

In [19]:
# save as csv
path = "Data/"

bwar_pitch.to_csv(path + 'bwar_pitch')

In [20]:
# load in csv
bwar_pitch = pd.read_csv('Data/bwar_pitch', index_col = 0)

In [21]:
# filter bwar_bat to 2000 - 2021
bwar_pitch_2021 = bwar_pitch[(bwar_pitch['year_ID'].between(2000, 2021))]
bwar_pitch_2021.sort_values(by = 'year_ID')

name_common   age  player_ID  year_ID team_ID  stint_ID lg_ID  \
48453        Ismael Valdez  26.0  valdeis01     2000     LAD         2    NL   
36746           Brad Penny  22.0  pennybr01     2000     FLA         1    NL   
15902          Aaron Fultz  26.0  fultzaa01     2000     SFG         1    NL   
36671           Jesus Pena  25.0   penaje01     2000     CHW         1    AL   
36670           Jesus Pena  25.0   penaje01     2000     BOS         2    AL   
...                    ...   ...        ...      ...     ...       ...   ...   
16240          Luis Garcia  24.0  garcilu05     2021     HOU         1    AL   
2190            Joe Barlow  25.0  barlojo01     2021     TEX         1    AL   
34785         Darien Núñez  28.0  nunezda01     2021     LAD         1    NL   
31583  Julian Merryweather  29.0  merryju01     2021     TOR         1    AL   
32901         Jovani Moran  24.0  moranjo01     2021     MIN         1    AL   

         salary  
48453       NaN  
36746  200000.0  
15902  200000.0  
36671       NaN  
36670       NaN  
...         ...  
16240       NaN  
2190        NaN  
34785       NaN  
31583       NaN  
32901       NaN  

[16555 rows x 8 columns]

In [22]:
bwar_pitch_2021.describe()

age       year_ID      stint_ID        salary
count  16555.000000  16555.000000  16555.000000  8.854000e+03
mean      28.208819   2011.200725      1.091755  3.114571e+06
std        4.092570      6.424128      0.303380  4.556342e+06
min       19.000000   2000.000000      1.000000  1.000000e+05
25%       25.000000   2006.000000      1.000000  4.935000e+05
50%       28.000000   2012.000000      1.000000  1.050000e+06
75%       31.000000   2017.000000      1.000000  4.000000e+06
max       49.000000   2021.000000      5.000000  4.000000e+07

In [23]:
# how many missing values in pitching salary column?
print("Missing bWAR Pitching Salary Data:")
bwar_pitch_2021['salary'].isna().value_counts(normalize = True)

Missing bWAR Pitching Salary Data:


False    0.534823
True     0.465177
Name: salary, dtype: float64

## Formatting the Batting Table

**Join `batting_2021` with `bwar_bat_2021` on the following columns:**

**`batting_2021`** 
- `Name`
- `Season`

**`bwar_bat_2021`**  
- `name_common`
- `year_ID`

In [24]:
# join batting_2021 onto bwar_bat_2021 (left join)
batting_merged = pd.merge(bwar_bat_2021, batting_2021, how='left',left_on=['name_common','year_ID'],right_on=['Name','Season'])

In [25]:
batting_merged.shape

(9956, 330)

In [26]:
batting_merged.head()

name_common    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID  \
0         Jeff Abbott  110010.0  abbotje01     2000     CHW         1    AL   
1         Kurt Abbott  110011.0  abbotku01     2000     NYM         1    NL   
2  Reggie Abercrombie  430631.0  abercre01     2006     FLA         1    NL   
3     Brent Abernathy  150466.0  abernbr01     2001     TBD         1    AL   
4     Brent Abernathy  150466.0  abernbr01     2002     TBD         1    AL   

  pitcher  G_x   PA_x    salary       IDfg  Season                Name Team  \
0       N   80  242.0  255000.0  1000011.0  2000.0         Jeff Abbott  CHW   
1       N   79  173.0  500000.0  1000013.0  2000.0         Kurt Abbott  NYM   
2       N  111  281.0  327000.0     2187.0  2006.0  Reggie Abercrombie  FLA   
3       N   79  335.0       NaN     1192.0  2001.0     Brent Abernathy  TBD   
4       N  117  504.0  215000.0     1192.0  2002.0     Brent Abernathy  TBD   

    Age    G_y     AB   PA_y      H    1B    2B   3B   HR     R   RBI    BB  \
0  27.0   80.0  215.0  242.0   59.0  40.0  15.0  1.0  3.0  31.0  29.0  21.0   
1  31.0   79.0  157.0  173.0   34.0  20.0   7.0  1.0  6.0  22.0  12.0  14.0   
2  24.0  111.0  255.0  281.0   54.0  35.0  12.0  2.0  5.0  39.0  24.0  18.0   
3  23.0   79.0  304.0  335.0   82.0  59.0  17.0  1.0  5.0  43.0  33.0  27.0   
4  24.0  117.0  463.0  504.0  112.0  88.0  18.0  4.0  2.0  46.0  40.0  25.0   

   IBB    SO  HBP   SF   SH  GDP    SB   CS    AVG     GB     FB    LD  IFFB  \
0  1.0  38.0  2.0  1.0  2.0  2.0   2.0  1.0  0.274    NaN    NaN   NaN   NaN   
1  2.0  51.0  1.0  1.0  0.0  2.0   1.0  1.0  0.217    NaN    NaN   NaN   NaN   
2  2.0  78.0  3.0  1.0  4.0  2.0   6.0  5.0  0.212   93.0   59.0  19.0   9.0   
3  1.0  35.0  0.0  1.0  3.0  3.0   8.0  3.0  0.270    NaN    NaN   NaN   NaN   
4  0.0  46.0  6.0  2.0  8.0  8.0  10.0  4.0  0.242  198.0  135.0  81.0  23.0   

   Pitches  Balls  Strikes   IFH    BU  BUH    BB%     K%  BB/K    OBP    SLG  \
0      NaN    NaN      NaN   NaN   NaN  NaN  0.087  0.157  0.55  0.343  0.395   
1      NaN    NaN      NaN   NaN   NaN  NaN  0.081  0.295  0.27  0.283  0.389   
2    990.0  307.0    683.0  11.0  11.0  5.0  0.064  0.278  0.23  0.271  0.333   
3      NaN    NaN      NaN   NaN   NaN  NaN  0.081  0.104  0.77  0.328  0.382   
4   1732.0  570.0   1162.0  12.0  13.0  4.0  0.050  0.091  0.54  0.288  0.311   

     OPS    ISO  BABIP   GB/FB    LD%    GB%    FB%  IFFB%  HR/FB   IFH%  \
0  0.738  0.121  0.320     NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1  0.672  0.172  0.277     NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2  0.604  0.122  0.283  0.0158  0.111  0.544  0.345  0.153  0.085  0.118   
3  0.710  0.112  0.291     NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4  0.599  0.069  0.264  0.0147  0.196  0.478  0.326  0.170  0.015  0.061   

    BUH%   wOBA  wRAA   wRC   Bat  Fld   Rep  Pos   RAR  WAR     Dol  Spd  \
0    NaN  0.328  -2.7  29.0  -4.2 -5.0   7.8 -1.6  -2.1 -0.2     NaN  3.9   
1    NaN  0.289  -7.8  15.0  -7.7 -2.0   5.6  1.5  -2.6 -0.3     NaN  4.2   
2  0.455  0.264 -16.3  19.0 -16.9 -5.6   8.8  0.9 -10.3 -1.0  ($4.8)  6.2   
3    NaN  0.312  -4.3  37.0  -5.0 -1.0  10.4  1.2   7.1  0.7     NaN  4.7   
4  0.308  0.269 -23.6  37.0 -24.7  3.3  15.4  1.6  -2.8 -0.3  ($1.1)  5.1   

   wRC+   WPA   -WPA  +WPA   RE24   REW   pLI  phLI    PH  WPA/LI  Clutch  \
0  87.0  0.37  -3.89  4.26  -6.74 -0.57  0.90  1.79  14.0   -0.16    0.57   
1  67.0 -0.40  -3.05  2.65 -12.42 -1.20  0.95  1.12  16.0   -0.88    0.46   
2  54.0 -1.40  -5.31  3.90 -12.00 -1.18  0.96  0.98  10.0   -0.94   -0.52   
3  88.0 -0.08  -5.86  5.77   2.30  0.23  0.96   NaN   0.0   -0.22    0.14   
4  60.0 -1.93 -10.36  8.44 -20.66 -2.02  1.12   NaN   0.0   -2.46    0.75   

   FB% (Pitch)   FBv    SL%   SLv    CT%   CTv    CB%   CBv    CH%   CHv  \
0          NaN   NaN    NaN   NaN    NaN   NaN    NaN   NaN    NaN   NaN   
1          NaN   NaN    NaN   NaN    NaN   NaN    NaN   NaN    NaN   NaN   
2     

Load in `positions` csv and merge with `batting_merged`

In [27]:
positions = pd.read_csv('Data/positions', index_col = 0)
positions.head()

playerID  yearID POS
106485  abbotje01    2000  OF
106486  abbotku01    2000  2B
106487  abbotku01    2000  3B
106488  abbotku01    2000  OF
106489  abbotku01    2000  SS

In [28]:
positions.shape

(40595, 3)

In [29]:
# join positions onto batting_merged (left join)
batting_merged = pd.merge(batting_merged, positions, 
                          how='left',
                          left_on=['player_ID','year_ID'],
                          right_on=['playerID','yearID'])

In [30]:
batting_merged.tail()

name_common    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID  \
18732  Mike Zunino  572287.0  zuninmi01     2016     SEA         1    AL   
18733  Mike Zunino  572287.0  zuninmi01     2017     SEA         1    AL   
18734  Mike Zunino  572287.0  zuninmi01     2018     SEA         1    AL   
18735  Mike Zunino  572287.0  zuninmi01     2019     TBR         1    AL   
18736  Mike Zunino  572287.0  zuninmi01     2021     TBR         1    AL   

      pitcher  G_x   PA_x     salary     IDfg  Season         Name Team   Age  \
18732       N   55  192.0        NaN  13265.0  2016.0  Mike Zunino  SEA  25.0   
18733       N  124  435.0   570000.0  13265.0  2017.0  Mike Zunino  SEA  26.0   
18734       N  113  405.0  2975000.0  13265.0  2018.0  Mike Zunino  SEA  27.0   
18735       N   90  289.0  4412000.0  13265.0  2019.0  Mike Zunino  TBR  28.0   
18736       N  109  375.0  2000000.0  13265.0  2021.0  Mike Zunino  TBR  30.0   

         G_y     AB   PA_y     H    1B    2B   3B    HR     R   RBI    BB  \
18732   55.0  164.0  192.0  34.0  15.0   7.0  0.0  12.0  16.0  31.0  21.0   
18733  124.0  387.0  435.0  97.0  47.0  25.0  0.0  25.0  52.0  64.0  39.0   
18734  113.0  373.0  405.0  75.0  37.0  18.0  0.0  20.0  37.0  44.0  24.0   
18735   90.0  266.0  289.0  44.0  24.0  10.0  1.0   9.0  30.0  32.0  20.0   
18736  109.0  333.0  375.0  72.0  26.0  11.0  2.0  33.0  64.0  62.0  34.0   

       IBB     SO  HBP   SF   SH  GDP   SB   CS    AVG    GB     FB    LD  \
18732  0.0   65.0  6.0  1.0  0.0  0.0  0.0  0.0  0.207  29.0   52.0  18.0   
18733  0.0  160.0  8.0  1.0  0.0  8.0  1.0  0.0  0.251  73.0  104.0  51.0   
18734  0.0  150.0  6.0  2.0  0.0  7.0  0.0  0.0  0.201  84.0   99.0  42.0   
18735  0.0   98.0  3.0  0.0  0.0  4.0  0.0  0.0  0.165  68.0   77.0  23.0   
18736  0.0  132.0  7.0  1.0  0.0  7.0  0.0  0.0  0.216  61.0  109.0  32.0   

       IFFB  Pitches  Balls  Strikes  IFH   BU  BUH    BB%     K%  BB/K  \
18732   6.0    810.0  313.0    497.0  3.0  1.0  0.0  0.109  0.339  0.32   
18733  13.0   1800.0  662.0   1138.0  3.0  0.0  0.0  0.090  0.368  0.24   
18734  13.0   1714.0  625.0   1089.0  6.0  0.0  0.0  0.059  0.370  0.16   
18735  14.0   1189.0  415.0    774.0  3.0  0.0  0.0  0.069  0.339  0.20   
18736  25.0   1502.0  566.0    936.0  5.0  0.0  0.0  0.091  0.352  0.26   

         OBP    SLG    OPS    ISO  BABIP   GB/FB    LD%    GB%    FB%  IFFB%  \
18732  0.318  0.470  0.787  0.262  0.250  0.0056  0.182  0.293  0.525  0.115   
18733  0.331  0.509  0.840  0.258  0.355  0.0070  0.224  0.320  0.456  0.125   
18734  0.259  0.410  0.669  0.209  0.268  0.0085  0.187  0.373  0.440  0.131   
18735  0.232  0.312  0.544  0.147  0.220  0.0088  0.137  0.405  0.458  0.182   
18736  0.301  0.559  0.860  0.342  0.231  0.0056  0.158  0.302  0.540  0.229   

       HR/FB   IFH%  BUH%   wOBA  wRAA   wRC   Bat   Fld   Rep  Pos   RAR  \
18732  0.231  0.103   0.0  0.338   3.1  26.0   3.8   2.3   5.8  3.6  15.4   
18733  0.240  0.041   0.0  0.355  12.4  65.0  13.8  11.2  13.4  8.3  46.2   
18734  0.202  0.071   0.0  0.288  -8.9  38.0  -8.3   9.9  12.1  7.8  20.8   
18735  0.117  0.044   0.0  0.235 -21.2  15.0 -19.9   6.8   9.1  5.9   3.9   
18736  0.303  0.082   0.0  0.358  13.7  59.0  15.5   9.0  11.7  7.4  45.1   

       WAR    Dol  Spd   wRC+   WPA  -WPA  +WPA   RE24   REW   pLI  phLI   PH  \
18732  1.6  $12.6  0.1  117.0  0.84 -3.12  3.96   8.57  0.91  1.00  2.78  3.0   
18733  4.6  $36.8  1.9  126.0  1.02 -7.83  8.85   6.94  0.67  0.95  1.81  4.0   
18734  2.1  $17.2  0.9   83.0 -1.31 -8.27  6.96 -18.61 -2.01  1.03  2.14  7.0   
18735  0.4   $3.0  2.9   46.0 -1.91 -5.94  4.03 -20.89 -2.00  0.99  1.82  2.0   
18736  4.5  $36.2  3.7  134.0  0.61 -6.92  7.54  10.89  1.13  0.99  3.02  5.0   

       WPA/LI  Clutch  FB% (Pitch)   FBv    SL%   SLv    CT%   CTv    CB%  \
18732    0.51    0.33        0.557  92.4  0.166  84.8  0.052  87.9  0.108   
18733    1.15   -0.07        0.581  92.9  0.176  84.4  0.062  88.4  0.079   
18734   -1.02   -0.25   

In [31]:
batting_merged.shape

(18737, 333)

**Filling missing `salary` values.**

We have missing salary values for approximately 25% of batters in the `batting_merged` dataset. Most of the missing values are likely due to clerical input whereby a player's salary for future years may not be filled in, but may be accounted for in an earlier year. 

For example, take a look at the player Reggie Abercrombie in the 2006 season. His 2006 salary is available, however, 2007 season is not. And in 2008, Reggie changed teams and his salary is still missing. Similarly, we see missing values for Brent Abernathy, whereby his 2001 and 2005 seasons contain missing salary data but his salary is available in the 2002 season. 

There are also players who were traded mid-season. The player is accounted for twice for a single season because they played for a different team and the salary may be missing for one of the entries. However, we will treat their salaries as the same for both teams. 

To fill in these missing values, we will assume the following:
- A player's salary is the same across consecutive seasons even if the salary is missing
- Players traded mid-season will have the same salary for that season across both teams


In [32]:
# make a copy and specify which column to fill missing values
salary_col = ['salary']
batting_filled = batting_merged.copy()

In [33]:
# forward and backward fills missing values grouped by name
batting_filled.loc[:,salary_col] = batting_merged.loc[:,salary_col].groupby(batting_filled['name_common']).ffill().bfill()

In [34]:
# check
batting_filled.tail(10)

name_common    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID  \
18727   Ben Zobrist  450314.0  zobribe01     2019     CHC         1    NL   
18728  Julio Zuleta  275936.0  zuletju01     2001     CHC         1    NL   
18729   Mike Zunino  572287.0  zuninmi01     2013     SEA         1    AL   
18730   Mike Zunino  572287.0  zuninmi01     2014     SEA         1    AL   
18731   Mike Zunino  572287.0  zuninmi01     2015     SEA         1    AL   
18732   Mike Zunino  572287.0  zuninmi01     2016     SEA         1    AL   
18733   Mike Zunino  572287.0  zuninmi01     2017     SEA         1    AL   
18734   Mike Zunino  572287.0  zuninmi01     2018     SEA         1    AL   
18735   Mike Zunino  572287.0  zuninmi01     2019     TBR         1    AL   
18736   Mike Zunino  572287.0  zuninmi01     2021     TBR         1    AL   

      pitcher  G_x   PA_x      salary       IDfg  Season          Name Team  \
18727       N   47  176.0  12500000.0     7435.0  2019.0   Ben Zobrist  CHC   
18728       N   49  118.0    200000.0  1014462.0  2001.0  Julio Zuleta  CHC   
18729       N   52  193.0    504100.0    13265.0  2013.0   Mike Zunino  SEA   
18730       N  131  476.0    504100.0    13265.0  2014.0   Mike Zunino  SEA   
18731       N  112  386.0    523500.0    13265.0  2015.0   Mike Zunino  SEA   
18732       N   55  192.0    523500.0    13265.0  2016.0   Mike Zunino  SEA   
18733       N  124  435.0    570000.0    13265.0  2017.0   Mike Zunino  SEA   
18734       N  113  405.0   2975000.0    13265.0  2018.0   Mike Zunino  SEA   
18735       N   90  289.0   4412000.0    13265.0  2019.0   Mike Zunino  TBR   
18736       N  109  375.0   2000000.0    13265.0  2021.0   Mike Zunino  TBR   

        Age    G_y     AB   PA_y     H    1B    2B   3B    HR     R   RBI  \
18727  38.0   47.0  150.0  176.0  39.0  33.0   5.0  0.0   1.0  24.0  17.0   
18728  26.0   49.0  106.0  118.0  23.0  14.0   3.0  0.0   6.0  11.0  24.0   
18729  22.0   52.0  173.0  193.0  37.0  27.0   5.0  0.0   5.0  22.0  14.0   
18730  23.0  131.0  438.0  476.0  87.0  43.0  20.0  2.0  22.0  51.0  60.0   
18731  24.0  112.0  350.0  386.0  61.0  39.0  11.0  0.0  11.0  28.0  28.0   
18732  25.0   55.0  164.0  192.0  34.0  15.0   7.0  0.0  12.0  16.0  31.0   
18733  26.0  124.0  387.0  435.0  97.0  47.0  25.0  0.0  25.0  52.0  64.0   
18734  27.0  113.0  373.0  405.0  75.0  37.0  18.0  0.0  20.0  37.0  44.0   
18735  28.0   90.0  266.0  289.0  44.0  24.0  10.0  1.0   9.0  30.0  32.0   
18736  30.0  109.0  333.0  375.0  72.0  26.0  11.0  2.0  33.0  64.0  62.0   

         BB  IBB     SO   HBP   SF   SH   GDP   SB   CS    AVG    GB     FB  \
18727  23.0  0.0   24.0   1.0  2.0  0.0   6.0  0.0  0.0  0.260  70.0   33.0   
18728   8.0  1.0   32.0   3.0  1.0  0.0   3.0  0.0  1.0  0.217   NaN    NaN   
18729  16.0  0.0   49.0   3.0  1.0  0.0   5.0  1.0  0.0  0.214  53.0   48.0   
18730  17.0  1.0  158.0  17.0  4.0  0.0  13.0  0.0  3.0  0.199  96.0  140.0   
18731  21.0  0.0  132.0   5.0  2.0  8.0   6.0  0.0  1.0  0.174  71.0  109.0   
18732  21.0  0.0   65.0   6.0  1.0  0.0   0.0  0.0  0.0  0.207  29.0   52.0   
18733  39.0  0.0  160.0   8.0  1.0  0.0   8.0  1.0  0.0  0.251  73.0  104.0   
18734  24.0  0.0  150.0   6.0  2.0  0.0   7.0  0.0  0.0  0.201  84.0   99.0   
18735  20.0  0.0   98.0   3.0  0.0  0.0   4.0  0.0  0.0  0.165  68.0   77.0   
18736  34.0  0.0  132.0   7.0  1.0  0.0   7.0  0.0  0.0  0.216  61.0  109.0   

         LD  IFFB  Pitches  Balls  Strikes  IFH    BU  BUH    BB%     K%  \
18727  24.0   4.0    710.0  299.0    411.0  3.0   1.0  1.0  0.131  0.136   
18728   NaN   NaN      NaN    NaN      NaN  NaN   NaN  NaN  0.068  0.271   
18729  23.0   7.0    767.0  283.0    484.0  5.0   1.0  0.0  0.083  0.254   
18730  47.0   8.0   1773.0  578.0   1195.0  4.0   1.0  0.0  0.036  0.332   
18731  38.0  16.0   1582.0  554.0   1028.0  3.0  10.0  0.0  0.054  0.342   
18732  18.0   6.0    810.0  313.0    497.0  3.0   1.0  0.0  0.109  0.339   
18733  51.0  13.0   1800.0  662.0   1138

In [35]:
# check
batting_filled.salary.isna().sum()

0

In [36]:
batting_filled.describe()

mlb_ID       year_ID      stint_ID           G_x          PA_x  \
count   18737.000000  18737.000000  18737.000000  18737.000000  18737.000000   
mean   377544.439611   2010.615147      1.101670     93.564605    331.975823   
std    182256.634054      6.372484      0.318893     38.225116    177.724853   
min    110010.000000   2000.000000      1.000000     22.000000    100.000000   
25%    150052.000000   2005.000000      1.000000     59.000000    179.000000   
50%    430948.000000   2011.000000      1.000000     91.000000    287.000000   
75%    516416.000000   2016.000000      1.000000    127.000000    467.000000   
max    683734.000000   2021.000000      4.000000    163.000000    778.000000   

             salary          IDfg        Season           Age           G_y  \
count  1.873700e+04  1.835600e+04  18356.000000  18356.000000  18356.000000   
mean   3.138487e+06  2.083374e+04   2010.602310     29.003868    103.819786   
std    4.544970e+06  1.243969e+05      6.363085      3.981274     36.070109   
min    1.000000e+05  1.000000e+00   2000.000000     19.000000     22.000000   
25%    4.950000e+05  1.058000e+03   2005.000000     26.000000     75.000000   
50%    1.000000e+06  3.255000e+03   2011.000000     29.000000    106.000000   
75%    4.000000e+06  8.623000e+03   2016.000000     32.000000    135.000000   
max    3.766667e+07  1.014462e+06   2021.000000     47.000000    163.000000   

                 AB          PA_y             H            1B            2B  \
count  18356.000000  18356.000000  18356.000000  18356.000000  18356.000000   
mean     328.826106    367.816572     85.906243     56.734147     17.176727   
std      157.942189    177.657385     47.164385     31.544783     10.485472   
min       46.000000     58.000000      6.000000      2.000000      0.000000   
25%      190.000000    211.000000     46.000000     30.000000      9.000000   
50%      309.000000    344.000000     78.000000     51.000000     15.000000   
75%      465.000000    518.000000    122.000000     80.000000     24.000000   
max      716.000000    778.000000    262.000000    225.000000     59.000000   

                 3B            HR             R           RBI            BB  \
count  18356.000000  18356.000000  18356.000000  18356.000000  18356.000000   
mean       1.768141     10.227228     44.057801     41.867945     31.084550   
std        2.052129      9.217051     26.430946     27.165780     21.353826   
min        0.000000      0.000000      2.000000      0.000000      1.000000   
25%        0.000000      3.000000     22.000000     20.000000     15.000000   
50%        1.000000      7.000000     38.000000     36.000000     26.000000   
75%        3.000000     14.000000     62.000000     58.000000     42.000000   
max       23.000000     73.000000    152.000000    160.000000    232.000000   

               IBB            SO           HBP            SF            SH  \
count  18356.00000  18356.000000  18356.000000  18356.000000  18356.000000   
mean       2.10073     67.646927      3.456145      2.623229      1.769394   
std        3.33381     36.184914      3.565794      2.250243      2.534823   
min        0.00000      3.000000      0.000000      0.000000      0.000000   
25%        0.00000     39.000000      1.000000      1.000000      0.000000   
50%        1.00000     61.000000      2.000000      2.000000      1.000000   
75%        3.00000     89.000000      5.000000      4.000000      3.000000   
max      120.00000    223.000000     31.000000     16.000000     24.000000   

                GDP            SB            CS           AVG            GB  \
count  18356.000000  18356.000000  18356.000000  18356.000000  16639.000000   
mean       7.500926      5.284975      2.166649      0.253511    113.282349   
std        5.192282      7.782792      2.583092      0.035540     60.618489   
min        0.000000      0.000000      0.000000      0.101000     11.000000   
25%        4.000000      1.000000      0.000000      0.2320

## Simplify and separate the `batting_filled` table

Because there are so many advanced statistics available in the dataframe, we can reduce the number of columns down to standard team statistics that do not include StatCast data. 

As mentioned earlier, StatCast was first introduced to MLB in 2014 and was not fully installed at every ballpark until 2015. Therfore, there will be many missing values for the advanced StatCast features. 

Thus, we will create two separate batting tables:
- `batting_basic` (containing basic data between 2000 - 2021)
- `batting_adv` (containing basic AND StatCast data between 2014 - 2021)

We'll also remove repetitive columns that we brought in during the initial merge of tables. 

In [37]:
batting_filled.head(1)

name_common    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID pitcher  \
0  Jeff Abbott  110010.0  abbotje01     2000     CHW         1    AL       N   

   G_x   PA_x    salary       IDfg  Season         Name Team   Age   G_y  \
0   80  242.0  255000.0  1000011.0  2000.0  Jeff Abbott  CHW  27.0  80.0   

      AB   PA_y     H    1B    2B   3B   HR     R   RBI    BB  IBB    SO  HBP  \
0  215.0  242.0  59.0  40.0  15.0  1.0  3.0  31.0  29.0  21.0  1.0  38.0  2.0   

    SF   SH  GDP   SB   CS    AVG  GB  FB  LD  IFFB  Pitches  Balls  Strikes  \
0  1.0  2.0  2.0  2.0  1.0  0.274 NaN NaN NaN   NaN      NaN    NaN      NaN   

   IFH  BU  BUH    BB%     K%  BB/K    OBP    SLG    OPS    ISO  BABIP  GB/FB  \
0  NaN NaN  NaN  0.087  0.157  0.55  0.343  0.395  0.738  0.121   0.32    NaN   

   LD%  GB%  FB%  IFFB%  HR/FB  IFH%  BUH%   wOBA  wRAA   wRC  Bat  Fld  Rep  \
0  NaN  NaN  NaN    NaN    NaN   NaN   NaN  0.328  -2.7  29.0 -4.2 -5.0  7.8   

   Pos  RAR  WAR  Dol  Spd  wRC+   WPA  -WPA  +WPA  RE24   REW  pLI  phLI  \
0 -1.6 -2.1 -0.2  NaN  3.9  87.0  0.37 -3.89  4.26 -6.74 -0.57  0.9  1.79   

     PH  WPA/LI  Clutch  FB% (Pitch)  FBv  SL%  SLv  CT%  CTv  CB%  CBv  CH%  \
0  14.0   -0.16    0.57          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   CHv  SF%  SFv  KN%  KNv  XX%  PO%  wFB  wSL  wCT  wCB  wCH  wSF  wKN  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   wFB/C  wSL/C  wCT/C  wCB/C  wCH/C  wSF/C  wKN/C  O-Swing%  Z-Swing%  \
0    NaN    NaN    NaN    NaN    NaN    NaN    NaN       NaN       NaN   

   Swing%  O-Contact%  Z-Contact%  Contact%  Zone%  F-Strike%  SwStr%  BsR  \
0     NaN         NaN         NaN       NaN    NaN        NaN     NaN  0.0   

   FA% (sc)  FT% (sc)  FC% (sc)  FS% (sc)  FO% (sc)  SI% (sc)  SL% (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   CU% (sc)  KC% (sc)  EP% (sc)  CH% (sc)  SC% (sc)  KN% (sc)  UN% (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   vFA (sc)  vFT (sc)  vFC (sc)  vFS (sc)  vFO (sc)  vSI (sc)  vSL (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   vCU (sc)  vKC (sc)  vEP (sc)  vCH (sc)  vSC (sc)  vKN (sc)  FA-X (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN        NaN   

   FT-X (sc)  FC-X (sc)  FS-X (sc)  FO-X (sc)  SI-X (sc)  SL-X (sc)  \
0        NaN        NaN        NaN        NaN        NaN        NaN   

   CU-X (sc)  KC-X (sc)  EP-X (sc)  CH-X (sc)  SC-X (sc)  KN-X (sc)  \
0        NaN        NaN        NaN        NaN        NaN        NaN   

   FA-Z (sc)  FT-Z (sc)  FC-Z (sc)  FS-Z (sc)  FO-Z (sc)  SI-Z (sc)  \
0        NaN        NaN        NaN        NaN        NaN        NaN   

   SL-Z (sc)  CU-Z (sc)  KC-Z (sc)  EP-Z (sc)  CH-Z (sc)  SC-Z (sc)  \
0        NaN        NaN        NaN        NaN        NaN        NaN   

   KN-Z (sc)  wFA (sc)  wFT (sc)  wFC (sc)  wFS (sc)  wFO (sc)  wSI (sc)  \
0        NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   wSL (sc)  wCU (sc)  wKC (sc)  wEP (sc)  wCH (sc)  wSC (sc)  wKN (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   wFA/C (sc)  wFT/C (sc)  wFC/C (sc)  wFS/C (sc)  wFO/C (sc)  wSI/C (sc)  \
0         NaN         NaN         NaN         NaN         NaN         NaN   

   wSL/C (sc)  wCU/C (sc)  wKC/C (sc)  wEP/C (sc)  wCH/C (sc)  wSC/C (sc)  \
0         NaN         NaN         NaN         NaN         NaN         NaN   

   wKN/C (sc)  O-Swing% (sc)  Z-Swing% (sc)  Swing% (sc)  O-Contact% (sc)  \
0         NaN            NaN            NaN          NaN              NaN   

   Z-Contact% (sc)  Contact% (sc)  Zone% (sc)  Pace  Def  wSB  UBR  Age Rng  \
0              NaN            NaN         NaN   NaN -6.6  0.0  NaN  27 - 27   

   Off   Lg  wGDP  Pull%  Cent%  Oppo%  Soft%  Med%  Hard%   TTO%  CH% (pi)  \
0 -4.2  0.9   NaN    NaN    NaN    NaN    NaN   NaN    NaN  0.256       NaN   

   

In [38]:
# get all column names
with pd.option_context('display.max_seq_items', None):
    print (batting_filled.columns)

Index(['name_common', 'mlb_ID', 'player_ID', 'year_ID', 'team_ID', 'stint_ID',
       'lg_ID', 'pitcher', 'G_x', 'PA_x', 'salary', 'IDfg', 'Season', 'Name',
       'Team', 'Age', 'G_y', 'AB', 'PA_y', 'H', '1B', '2B', '3B', 'HR', 'R',
       'RBI', 'BB', 'IBB', 'SO', 'HBP', 'SF', 'SH', 'GDP', 'SB', 'CS', 'AVG',
       'GB', 'FB', 'LD', 'IFFB', 'Pitches', 'Balls', 'Strikes', 'IFH', 'BU',
       'BUH', 'BB%', 'K%', 'BB/K', 'OBP', 'SLG', 'OPS', 'ISO', 'BABIP',
       'GB/FB', 'LD%', 'GB%', 'FB%', 'IFFB%', 'HR/FB', 'IFH%', 'BUH%', 'wOBA',
       'wRAA', 'wRC', 'Bat', 'Fld', 'Rep', 'Pos', 'RAR', 'WAR', 'Dol', 'Spd',
       'wRC+', 'WPA', '-WPA', '+WPA', 'RE24', 'REW', 'pLI', 'phLI', 'PH',
       'WPA/LI', 'Clutch', 'FB% (Pitch)', 'FBv', 'SL%', 'SLv', 'CT%', 'CTv',
       'CB%', 'CBv', 'CH%', 'CHv', 'SF%', 'SFv', 'KN%', 'KNv', 'XX%', 'PO%',
       'wFB', 'wSL', 'wCT', 'wCB', 'wCH', 'wSF', 'wKN', 'wFB/C', 'wSL/C',
       'wCT/C', 'wCB/C', 'wCH/C', 'wSF/C', 'wKN/C', 'O-Swing%', 'Z-Swing%',
    

**Specify columns for `batting_basic`**

In [39]:
batting_filled.loc[8:15]

name_common    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID pitcher  \
8   Bobby Abreu  110029.0  abreubo01     2000     PHI         1    NL       N   
9   Bobby Abreu  110029.0  abreubo01     2001     PHI         1    NL       N   
10  Bobby Abreu  110029.0  abreubo01     2002     PHI         1    NL       N   
11  Bobby Abreu  110029.0  abreubo01     2003     PHI         1    NL       N   
12  Bobby Abreu  110029.0  abreubo01     2004     PHI         1    NL       N   
13  Bobby Abreu  110029.0  abreubo01     2005     PHI         1    NL       N   
14  Bobby Abreu  110029.0  abreubo01     2006     NYY         2    AL       N   
15  Bobby Abreu  110029.0  abreubo01     2006     NYY         2    AL       N   

    G_x   PA_x      salary   IDfg  Season         Name   Team   Age    G_y  \
8   154  680.0   2933333.0  945.0  2000.0  Bobby Abreu    PHI  26.0  154.0   
9   162  704.0   4983000.0  945.0  2001.0  Bobby Abreu    PHI  27.0  162.0   
10  157  685.0   6333333.0  945.0  2002.0  Bobby Abreu    PHI  28.0  157.0   
11  158  695.0   9100000.0  945.0  2003.0  Bobby Abreu    PHI  29.0  158.0   
12  159  713.0  10600000.0  945.0  2004.0  Bobby Abreu    PHI  30.0  159.0   
13  162  719.0  13100000.0  945.0  2005.0  Bobby Abreu    PHI  31.0  162.0   
14   58  248.0  13100000.0  945.0  2006.0  Bobby Abreu  - - -  32.0  156.0   
15   58  248.0  13100000.0  945.0  2006.0  Bobby Abreu  - - -  32.0  156.0   

       AB   PA_y      H     1B    2B    3B    HR      R    RBI     BB   IBB  \
8   576.0  680.0  182.0  105.0  42.0  10.0  25.0  103.0   79.0  100.0   9.0   
9   588.0  704.0  170.0   87.0  48.0   4.0  31.0  118.0  110.0  106.0  11.0   
10  572.0  685.0  176.0  100.0  50.0   6.0  20.0  102.0   85.0  104.0   9.0   
11  577.0  695.0  173.0  117.0  35.0   1.0  20.0   99.0  101.0  109.0  13.0   
12  574.0  713.0  173.0   95.0  47.0   1.0  30.0  118.0  105.0  127.0  10.0   
13  588.0  719.0  168.0  106.0  37.0   1.0  24.0  104.0  102.0  117.0  15.0   
14  548.0  686.0  163.0  105.0  41.0   2.0  15.0   98.0  107.0  124.0   6.0   
15  548.0  686.0  163.0  105.0  41.0   2.0  15.0   98.0  107.0  124.0   6.0   

       SO  HBP   SF   SH   GDP    SB    CS    AVG     GB     FB     LD  IFFB  \
8   116.0  1.0  3.0  0.0  12.0  28.0   8.0  0.316    NaN    NaN    NaN   NaN   
9   137.0  1.0  9.0  0.0  13.0  36.0  14.0  0.289    NaN    NaN    NaN   NaN   
10  117.0  3.0  6.0  0.0  11.0  31.0  12.0  0.308  197.0  151.0  109.0   6.0   
11  126.0  2.0  7.0  0.0  13.0  22.0   9.0  0.300  225.0  122.0  110.0   2.0   
12  116.0  5.0  7.0  0.0   5.0  40.0   5.0  0.301  196.0  160.0  109.0   7.0   
13  134.0  6.0  8.0  0.0   7.0  31.0   9.0  0.286  216.0  133.0  111.0   5.0   
14  138.0  3.0  9.0  2.0  13.0  30.0   6.0  0.297  188.0  121.0  110.0   1.0   
15  138.0  3.0  9.0  2.0  13.0  30.0   6.0  0.297  188.0  121.0  110.0   1.0   

    Pitches   Balls  Strikes   IFH   BU  BUH    BB%     K%  BB/K    OBP  \
8       NaN     NaN      NaN   NaN  NaN  NaN  0.147  0.171  0.86  0.416   
9       NaN     NaN      NaN   NaN  NaN  NaN  0.151  0.195  0.77  0.393   
10   2960.0  1251.0   1709.0   8.0  4.0  3.0  0.152  0.171  0.89  0.413   
11   2999.0  1277.0   1722.0  16.0  1.0  1.0  0.157  0.181  0.87  0.409   
12   3081.0  1339.0   1742.0  11.0  0.0  0.0  0.178  0.163  1.09  0.428   
13   3164.0  1369.0   1795.0  10.0  2.0  2.0  0.163  0.186  0.87  0.405   
14   3059.0  1319.0   1740.0  10.0  2.0  0.0  0.181  0.201  0.90  0.424   
15   3059.0  1319.0   1740.0  10.0  2.0  0.0  0.181  0.201  0.90  0.424   

      SLG    OPS    ISO  BABIP   GB/FB    LD%    GB%    FB%  IFFB%  HR/FB  \
8   0.554  0.970  0.238  0.358     NaN    NaN    NaN    NaN    NaN    NaN   
9   0.543  0.936  0.253  0.324     NaN    NaN    NaN    NaN    NaN    NaN   
10  0.521  0.934  0.213  0.354  0.0130  0.239  0.431  0.330  0.040  0.132   
11  0.468  0.877  0.168  0.349  0.0184  0.241  0.492  0.267  0.016  0.164   
12  0.544  0.971  0.242  0.329  0.0123  0.234  0.422  0.344  0.044  0.188   


In [40]:
# basic batting columns to keep
batting_basic_columns = ['Season', 'Name', 'POS','team_ID', 'Age', 'salary', 'G_y', 'AB', 'PA_y', 
                         'AVG', 'H', '1B', '2B', '3B', 'HR', 'R', 'RBI', 'BB', 'IBB', 
                         'SO', 'HBP', 'SF', 'SH', 'GDP', 'SB', 'CS', 'OBP', 'SLG', 'OPS']

# create new batting_basic table with specified columns i want
batting_basic = batting_filled[batting_basic_columns]

# renaming columns
batting_basic = batting_basic.rename(columns = {'G_y':'G',
                                                'team_ID': 'Team',
                                                'PA_y':'PA',
                                                'salary':'Salary'})

**Specify columns for `batting_adv`**

In [41]:
# adv batting columns to keep
batting_adv_columns = ['Season', 'Name', 'POS', 'team_ID', 'Age', 'salary', 
                       'G_y', 'AB', 'PA_y', 'H', '1B', '2B', '3B', 'HR', 'R',
                       'RBI', 'BB', 'IBB', 'SO', 'HBP', 'SF', 'SH', 'GDP', 'SB', 'CS', 'AVG',
                       'GB', 'FB', 'LD', 'IFFB', 'Pitches', 'Balls', 'Strikes', 'IFH', 'BU',
                       'BUH', 'BB%', 'K%', 'BB/K', 'OBP', 'SLG', 'OPS', 'ISO', 'BABIP',
                       'GB/FB', 'LD%', 'GB%', 'FB%', 'IFFB%', 'HR/FB', 'IFH%', 'BUH%', 'wOBA',
                       'wRAA', 'wRC', 'Bat', 'Fld', 'Rep', 'Pos', 'RAR', 'WAR', 'Dol', 'Spd',
                       'wRC+', 'WPA', '-WPA', '+WPA', 'RE24', 'REW', 'pLI', 'phLI', 'PH',
                       'WPA/LI', 'Clutch', 'FB% (Pitch)', 'FBv', 'SL%', 'SLv', 'CT%', 'CTv',
                       'CB%', 'CBv', 'CH%', 'CHv', 'SF%', 'SFv', 'KN%', 'KNv', 'XX%', 'PO%',
                       'wFB', 'wSL', 'wCT', 'wCB', 'wCH', 'wSF', 'wKN', 'wFB/C', 'wSL/C',
                       'wCT/C', 'wCB/C', 'wCH/C', 'wSF/C', 'wKN/C', 'O-Swing%', 'Z-Swing%',
                       'Swing%', 'O-Contact%', 'Z-Contact%', 'Contact%', 'Zone%', 'F-Strike%',
                       'SwStr%', 'BsR', 'FA% (sc)', 'FT% (sc)', 'FC% (sc)', 'FS% (sc)',
                       'FO% (sc)', 'SI% (sc)', 'SL% (sc)', 'CU% (sc)', 'KC% (sc)', 'EP% (sc)',
                       'CH% (sc)', 'SC% (sc)', 'KN% (sc)', 'UN% (sc)', 'vFA (sc)', 'vFT (sc)',
                       'vFC (sc)', 'vFS (sc)', 'vFO (sc)', 'vSI (sc)', 'vSL (sc)', 'vCU (sc)',
                       'vKC (sc)', 'vEP (sc)', 'vCH (sc)', 'vSC (sc)', 'vKN (sc)', 'FA-X (sc)',
                       'FT-X (sc)', 'FC-X (sc)', 'FS-X (sc)', 'FO-X (sc)', 'SI-X (sc)',
                       'SL-X (sc)', 'CU-X (sc)', 'KC-X (sc)', 'EP-X (sc)', 'CH-X (sc)',
                       'SC-X (sc)', 'KN-X (sc)', 'FA-Z (sc)', 'FT-Z (sc)', 'FC-Z (sc)',
                       'FS-Z (sc)', 'FO-Z (sc)', 'SI-Z (sc)', 'SL-Z (sc)', 'CU-Z (sc)',
                       'KC-Z (sc)', 'EP-Z (sc)', 'CH-Z (sc)', 'SC-Z (sc)', 'KN-Z (sc)',
                       'wFA (sc)', 'wFT (sc)', 'wFC (sc)', 'wFS (sc)', 'wFO (sc)', 'wSI (sc)',
                       'wSL (sc)', 'wCU (sc)', 'wKC (sc)', 'wEP (sc)', 'wCH (sc)', 'wSC (sc)',
                       'wKN (sc)', 'wFA/C (sc)', 'wFT/C (sc)', 'wFC/C (sc)', 'wFS/C (sc)',
                       'wFO/C (sc)', 'wSI/C (sc)', 'wSL/C (sc)', 'wCU/C (sc)', 'wKC/C (sc)',
                       'wEP/C (sc)', 'wCH/C (sc)', 'wSC/C (sc)', 'wKN/C (sc)', 'O-Swing% (sc)',
                       'Z-Swing% (sc)', 'Swing% (sc)', 'O-Contact% (sc)', 'Z-Contact% (sc)',
                       'Contact% (sc)', 'Zone% (sc)', 'Pace', 'Def', 'wSB', 'UBR', 'Age Rng',
                       'Off', 'Lg', 'wGDP', 'Pull%', 'Cent%', 'Oppo%', 'Soft%', 'Med%',
                       'Hard%', 'TTO%', 'CH% (pi)', 'CS% (pi)', 'CU% (pi)', 'FA% (pi)',
                       'FC% (pi)', 'FS% (pi)', 'KN% (pi)', 'SB% (pi)', 'SI% (pi)', 'SL% (pi)',
                       'XX% (pi)', 'vCH (pi)', 'vCS (pi)', 'vCU (pi)', 'vFA (pi)', 'vFC (pi)',
                       'vFS (pi)', 'vKN (pi)', 'vSB (pi)', 'vSI (pi)', 'vSL (pi)', 'vXX (pi)',
                       'CH-X (pi)', 'CS-X (pi)', 'CU-X (pi)', 'FA-X (pi)', 'FC-X (pi)',
                       'FS-X (pi)', 'KN-X (pi)', 'SB-X (pi)', 'SI-X (pi)', 'SL-X (pi)',
                       'XX-X (pi)', 'CH-Z (pi)', 'CS-Z (pi)', 'CU-Z (pi)', 'FA-Z (pi)',
                       'FC-Z (pi)', 'FS-Z (pi)', 'KN-Z (pi)', 'SB-Z (pi)', 'SI-Z (pi)',
                       'SL-Z (pi)', 'XX-Z (pi)', 'wCH (pi)', 'wCS (pi)', 'wCU (pi)',
                       'wFA (pi)', 'wFC (pi)', 'wFS (pi)', 'wKN (pi)', 'wSB (pi)', 'wSI (pi)',
                       'wSL (pi)', 'wXX (pi)', 'wCH/C (pi)', 'wCS/C (pi)', 'wCU/C (pi)',
                       'wFA/C (pi)', 'wFC/C (pi)', 'wFS/C (pi)', 'wKN/C (pi)', 'wSB/C (pi)',
                       'wSI/C (pi)', 'wSL/C (pi)', 'wXX/C (pi)', 'O-Swing% (pi)',
                       'Z-Swing% (pi)', 'Swing% (pi)', 'O-Contact% (pi)', 'Z-Contact% (pi)',
                       'Contact% (pi)', 'Zone% (pi)', 'Pace (pi)', 'FRM', 'AVG+', 'BB%+',
                       'K%+', 'OBP+', 'SLG+', 'ISO+', 'BABIP+', 'LD+%', 'GB%+', 'FB%+',
                       'HR/FB%+', 'Pull%+', 'Cent%+', 'Oppo%+', 'Soft%+', 'Med%+', 'Hard%+',
                       'EV', 'LA', 'Barrels', 'Barrel%', 'maxEV', 'HardHit', 'HardHit%',
                       'Events', 'CStr%', 'CSW%', 'xBA', 'xSLG', 'xwOBA']

# create new batting_adv table with specified columns i want
batting_adv = batting_filled[batting_adv_columns]

# renaming columns
batting_adv = batting_adv.rename(columns = {'G_y':'G',
                                            'team_ID': 'Team',
                                            'PA_y':'PA',
                                            'salary':'Salary'})

**Checking and cleaning `batting_basic` first:**

In [42]:
# check
batting_basic.head()

Season         Name POS Team   Age    Salary     G     AB     PA    AVG  \
0  2000.0  Jeff Abbott  OF  CHW  27.0  255000.0  80.0  215.0  242.0  0.274   
1  2000.0  Kurt Abbott  2B  NYM  31.0  500000.0  79.0  157.0  173.0  0.217   
2  2000.0  Kurt Abbott  3B  NYM  31.0  500000.0  79.0  157.0  173.0  0.217   
3  2000.0  Kurt Abbott  OF  NYM  31.0  500000.0  79.0  157.0  173.0  0.217   
4  2000.0  Kurt Abbott  SS  NYM  31.0  500000.0  79.0  157.0  173.0  0.217   

      H    1B    2B   3B   HR     R   RBI    BB  IBB    SO  HBP   SF   SH  \
0  59.0  40.0  15.0  1.0  3.0  31.0  29.0  21.0  1.0  38.0  2.0  1.0  2.0   
1  34.0  20.0   7.0  1.0  6.0  22.0  12.0  14.0  2.0  51.0  1.0  1.0  0.0   
2  34.0  20.0   7.0  1.0  6.0  22.0  12.0  14.0  2.0  51.0  1.0  1.0  0.0   
3  34.0  20.0   7.0  1.0  6.0  22.0  12.0  14.0  2.0  51.0  1.0  1.0  0.0   
4  34.0  20.0   7.0  1.0  6.0  22.0  12.0  14.0  2.0  51.0  1.0  1.0  0.0   

   GDP   SB   CS    OBP    SLG    OPS  
0  2.0  2.0  1.0  0.343  0.395  0.738  
1  2.0  1.0  1.0  0.283  0.389  0.672  
2  2.0  1.0  1.0  0.283  0.389  0.672  
3  2.0  1.0  1.0  0.283  0.389  0.672  
4  2.0  1.0  1.0  0.283  0.389  0.672

In [43]:
batting_basic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18737 entries, 0 to 18736
Data columns (total 29 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Season  18356 non-null  float64
 1   Name    18356 non-null  object 
 2   POS     18603 non-null  object 
 3   Team    18737 non-null  object 
 4   Age     18356 non-null  float64
 5   Salary  18737 non-null  float64
 6   G       18356 non-null  float64
 7   AB      18356 non-null  float64
 8   PA      18356 non-null  float64
 9   AVG     18356 non-null  float64
 10  H       18356 non-null  float64
 11  1B      18356 non-null  float64
 12  2B      18356 non-null  float64
 13  3B      18356 non-null  float64
 14  HR      18356 non-null  float64
 15  R       18356 non-null  float64
 16  RBI     18356 non-null  float64
 17  BB      18356 non-null  float64
 18  IBB     18356 non-null  float64
 19  SO      18356 non-null  float64
 20  HBP     18356 non-null  float64
 21  SF      18356 non-null  float64
 22

In [44]:
# check missing values
batting_basic.isna().sum()

Season    381
Name      381
POS       134
Team        0
Age       381
Salary      0
G         381
AB        381
PA        381
AVG       381
H         381
1B        381
2B        381
3B        381
HR        381
R         381
RBI       381
BB        381
IBB       381
SO        381
HBP       381
SF        381
SH        381
GDP       381
SB        381
CS        381
OBP       381
SLG       381
OPS       381
dtype: int64

**Handling missing values in `batting_basic`**

There are 381 missing values. 

There are missing values for Name and Age and may be attributed to the merging done earlier. 

We'll just drop these missing rows from `batting_basic`. 

In [45]:
# drop all missing nan rows
batting_basic = batting_basic.dropna()

In [46]:
# check
batting_basic.isna().sum()

Season    0
Name      0
POS       0
Team      0
Age       0
Salary    0
G         0
AB        0
PA        0
AVG       0
H         0
1B        0
2B        0
3B        0
HR        0
R         0
RBI       0
BB        0
IBB       0
SO        0
HBP       0
SF        0
SH        0
GDP       0
SB        0
CS        0
OBP       0
SLG       0
OPS       0
dtype: int64

**Removing duplicate entries**

In cases where a player was traded mid-season, there are entries where that player's cumulative statistics for one season is repeated twice for two teams. 

We will use the following method to remove these duplicate items from the dataframes moving forward:

In [47]:
# check for duplicates, for example Carlos Beltran played for the Yankees and Rangers in 2016
beltran = batting_basic[batting_basic['Name'] == 'Carlos Beltran']
beltran.tail()

Season            Name POS Team   Age      Salary      G     AB     PA  \
1306  2016.0  Carlos Beltran  OF  NYY  39.0  15000000.0  151.0  552.0  593.0   
1307  2016.0  Carlos Beltran  OF  NYY  39.0  15000000.0  151.0  552.0  593.0   
1308  2016.0  Carlos Beltran  OF  TEX  39.0  15000000.0  151.0  552.0  593.0   
1309  2016.0  Carlos Beltran  OF  TEX  39.0  15000000.0  151.0  552.0  593.0   
1310  2017.0  Carlos Beltran  OF  HOU  40.0  16000000.0  129.0  467.0  509.0   

        AVG      H     1B    2B   3B    HR     R   RBI    BB  IBB     SO  HBP  \
1306  0.295  163.0  101.0  33.0  0.0  29.0  73.0  93.0  35.0  4.0  101.0  2.0   
1307  0.295  163.0  101.0  33.0  0.0  29.0  73.0  93.0  35.0  4.0  101.0  2.0   
1308  0.295  163.0  101.0  33.0  0.0  29.0  73.0  93.0  35.0  4.0  101.0  2.0   
1309  0.295  163.0  101.0  33.0  0.0  29.0  73.0  93.0  35.0  4.0  101.0  2.0   
1310  0.231  108.0   65.0  29.0  0.0  14.0  60.0  51.0  33.0  3.0  102.0  3.0   

       SF   SH   GDP   SB   CS    OBP    SLG    OPS  
1306  4.0  0.0  19.0  1.0  0.0  0.337  0.513  0.850  
1307  4.0  0.0  19.0  1.0  0.0  0.337  0.513  0.850  
1308  4.0  0.0  19.0  1.0  0.0  0.337  0.513  0.850  
1309  4.0  0.0  19.0  1.0  0.0  0.337  0.513  0.850  
1310  6.0  0.0   9.0  0.0  0.0  0.283  0.383  0.666

In [48]:
# remove repeating total statistics for one season for traded players, keep first entry
batting_basic = batting_basic.drop_duplicates(['Name', 'Season'], keep = 'first')

In [49]:
# check that repeating Beltran in 2016 is fixed
batting_basic[batting_basic['Name'] == 'Carlos Beltran'].tail()

Season            Name POS Team   Age      Salary      G     AB     PA  \
1302  2013.0  Carlos Beltran  OF  STL  36.0  13000000.0  145.0  554.0  600.0   
1303  2014.0  Carlos Beltran  1B  NYY  37.0  15000000.0  109.0  403.0  449.0   
1305  2015.0  Carlos Beltran  OF  NYY  38.0  15000000.0  133.0  478.0  531.0   
1306  2016.0  Carlos Beltran  OF  NYY  39.0  15000000.0  151.0  552.0  593.0   
1310  2017.0  Carlos Beltran  OF  HOU  40.0  16000000.0  129.0  467.0  509.0   

        AVG      H     1B    2B   3B    HR     R   RBI    BB  IBB     SO  HBP  \
1302  0.296  164.0  107.0  30.0  3.0  24.0  79.0  84.0  38.0  1.0   90.0  1.0   
1303  0.233   94.0   56.0  23.0  0.0  15.0  46.0  49.0  37.0  2.0   80.0  4.0   
1305  0.276  132.0   78.0  34.0  1.0  19.0  57.0  67.0  45.0  2.0   85.0  2.0   
1306  0.295  163.0  101.0  33.0  0.0  29.0  73.0  93.0  35.0  4.0  101.0  2.0   
1310  0.231  108.0   65.0  29.0  0.0  14.0  60.0  51.0  33.0  3.0  102.0  3.0   

       SF   SH   GDP   SB   CS    OBP    SLG    OPS  
1302  6.0  1.0  12.0  2.0  1.0  0.339  0.491  0.830  
1303  5.0  0.0  11.0  3.0  1.0  0.301  0.402  0.703  
1305  6.0  0.0  12.0  0.0  0.0  0.337  0.471  0.808  
1306  4.0  0.0  19.0  1.0  0.0  0.337  0.513  0.850  
1310  6.0  0.0   9.0  0.0  0.0  0.283  0.383  0.666

**Correcting Duplicate Team Names**

When checking the unique values of teams, we find that there are more than the current 30 MLB teams. This is due to teams having changed their name abreviations over the years. 

We'll fix this so that there are only 30 unique team names. 

In [50]:
# check current unique Team IDs
display(batting_basic['Team'].unique())
len(batting_basic['Team'].unique())

array(['CHW', 'NYM', 'FLA', 'TBD', 'PHI', 'NYY', 'LAA', 'LAD', 'ARI',
       'SFG', 'SEA', 'ATL', 'COL', 'TBR', 'MIL', 'STL', 'WSN', 'TOR',
       'TEX', 'DET', 'MIN', 'MIA', 'CIN', 'BAL', 'KCR', 'CHC', 'BOS',
       'SDP', 'PIT', 'OAK', 'CLE', 'HOU', 'ANA', 'MON'], dtype=object)

34

In [51]:
# team names to replace 
replacement_names = {'ANA':'LAA', 'FLA':'MIA', 'MON':'WSN','TBD':'TBR'}

# replace the above names in Team
batting_basic['Team'] = batting_basic['Team'].replace(replacement_names)

In [52]:
display(batting_basic['Team'].unique())
len(batting_basic['Team'].unique())

array(['CHW', 'NYM', 'MIA', 'TBR', 'PHI', 'NYY', 'LAA', 'LAD', 'ARI',
       'SFG', 'SEA', 'ATL', 'COL', 'MIL', 'STL', 'WSN', 'TOR', 'TEX',
       'DET', 'MIN', 'CIN', 'BAL', 'KCR', 'CHC', 'BOS', 'SDP', 'PIT',
       'OAK', 'CLE', 'HOU'], dtype=object)

30

In [53]:
batting_basic.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9139 entries, 0 to 18736
Data columns (total 29 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Season  9139 non-null   float64
 1   Name    9139 non-null   object 
 2   POS     9139 non-null   object 
 3   Team    9139 non-null   object 
 4   Age     9139 non-null   float64
 5   Salary  9139 non-null   float64
 6   G       9139 non-null   float64
 7   AB      9139 non-null   float64
 8   PA      9139 non-null   float64
 9   AVG     9139 non-null   float64
 10  H       9139 non-null   float64
 11  1B      9139 non-null   float64
 12  2B      9139 non-null   float64
 13  3B      9139 non-null   float64
 14  HR      9139 non-null   float64
 15  R       9139 non-null   float64
 16  RBI     9139 non-null   float64
 17  BB      9139 non-null   float64
 18  IBB     9139 non-null   float64
 19  SO      9139 non-null   float64
 20  HBP     9139 non-null   float64
 21  SF      9139 non-null   float64
 22 

In [54]:
batting_basic.shape

(9139, 29)

**Adjusting for Inflation in 2021**

In [55]:
# import cpis
cpi = pd.read_csv('Data/CPI (1960 - 2022).csv')

In [56]:
# clean the cpi csv
cpi = cpi[:62]
cpi = cpi[['Year', 'Jan']]
cpi = cpi.rename({'Jan': 'CPI'}, axis = 'columns')

In [57]:
cpi

Year      CPI
0   1960.0   29.370
1   1961.0   29.840
2   1962.0   30.040
3   1963.0   30.440
4   1964.0   30.940
..     ...      ...
57  2017.0  243.618
58  2018.0  248.743
59  2019.0  252.470
60  2020.0  258.682
61  2021.0  262.200

[62 rows x 2 columns]

In [58]:
# join batting basic to cpi
batting_basic = batting_basic.merge(cpi, left_on='Season', right_on='Year', how='left')

In [59]:
# create new adjusted 2021 salary using 2021 CPI to determine specific year salary
batting_basic['Adjusted 2021 Salary'] = round((batting_basic['Salary'] * 262.2) / batting_basic['CPI'],0)

In [60]:
batting_basic = batting_basic[['Season', 'Name', 'POS', 'Team', 'Age', 'Salary', 'Adjusted 2021 Salary', 
                               'G', 'AB', 'PA', 'AVG', 'H', '1B', '2B', '3B', 'HR', 'R', 'RBI', 'BB', 
                               'IBB', 'SO', 'HBP', 'SF', 'SH', 'GDP', 'SB', 'CS', 'OBP', 'SLG', 'OPS']]

In [61]:
batting_basic.head()

Season                Name POS Team   Age    Salary  Adjusted 2021 Salary  \
0  2000.0         Jeff Abbott  OF  CHW  27.0  255000.0              394926.0   
1  2000.0         Kurt Abbott  2B  NYM  31.0  500000.0              774365.0   
2  2006.0  Reggie Abercrombie  OF  MIA  24.0  327000.0              430203.0   
3  2001.0     Brent Abernathy  2B  TBR  23.0  215000.0              321031.0   
4  2002.0     Brent Abernathy  2B  TBR  24.0  215000.0              317237.0   

       G     AB     PA    AVG      H    1B    2B   3B   HR     R   RBI    BB  \
0   80.0  215.0  242.0  0.274   59.0  40.0  15.0  1.0  3.0  31.0  29.0  21.0   
1   79.0  157.0  173.0  0.217   34.0  20.0   7.0  1.0  6.0  22.0  12.0  14.0   
2  111.0  255.0  281.0  0.212   54.0  35.0  12.0  2.0  5.0  39.0  24.0  18.0   
3   79.0  304.0  335.0  0.270   82.0  59.0  17.0  1.0  5.0  43.0  33.0  27.0   
4  117.0  463.0  504.0  0.242  112.0  88.0  18.0  4.0  2.0  46.0  40.0  25.0   

   IBB    SO  HBP   SF   SH  GDP    SB   CS    OBP    SLG    OPS  
0  1.0  38.0  2.0  1.0  2.0  2.0   2.0  1.0  0.343  0.395  0.738  
1  2.0  51.0  1.0  1.0  0.0  2.0   1.0  1.0  0.283  0.389  0.672  
2  2.0  78.0  3.0  1.0  4.0  2.0   6.0  5.0  0.271  0.333  0.604  
3  1.0  35.0  0.0  1.0  3.0  3.0   8.0  3.0  0.328  0.382  0.710  
4  0.0  46.0  6.0  2.0  8.0  8.0  10.0  4.0  0.288  0.311  0.599

**Save `batting_basic` as a csv**

In [62]:
path = "Data/"

batting_basic.to_csv(path + 'batting_basic')

**Checking and cleaning `batting_adv`:**

In [63]:
batting_adv.head(3)

Season         Name POS Team   Age    Salary     G     AB     PA     H  \
0  2000.0  Jeff Abbott  OF  CHW  27.0  255000.0  80.0  215.0  242.0  59.0   
1  2000.0  Kurt Abbott  2B  NYM  31.0  500000.0  79.0  157.0  173.0  34.0   
2  2000.0  Kurt Abbott  3B  NYM  31.0  500000.0  79.0  157.0  173.0  34.0   

     1B    2B   3B   HR     R   RBI    BB  IBB    SO  HBP   SF   SH  GDP   SB  \
0  40.0  15.0  1.0  3.0  31.0  29.0  21.0  1.0  38.0  2.0  1.0  2.0  2.0  2.0   
1  20.0   7.0  1.0  6.0  22.0  12.0  14.0  2.0  51.0  1.0  1.0  0.0  2.0  1.0   
2  20.0   7.0  1.0  6.0  22.0  12.0  14.0  2.0  51.0  1.0  1.0  0.0  2.0  1.0   

    CS    AVG  GB  FB  LD  IFFB  Pitches  Balls  Strikes  IFH  BU  BUH    BB%  \
0  1.0  0.274 NaN NaN NaN   NaN      NaN    NaN      NaN  NaN NaN  NaN  0.087   
1  1.0  0.217 NaN NaN NaN   NaN      NaN    NaN      NaN  NaN NaN  NaN  0.081   
2  1.0  0.217 NaN NaN NaN   NaN      NaN    NaN      NaN  NaN NaN  NaN  0.081   

      K%  BB/K    OBP    SLG    OPS    ISO  BABIP  GB/FB  LD%  GB%  FB%  \
0  0.157  0.55  0.343  0.395  0.738  0.121  0.320    NaN  NaN  NaN  NaN   
1  0.295  0.27  0.283  0.389  0.672  0.172  0.277    NaN  NaN  NaN  NaN   
2  0.295  0.27  0.283  0.389  0.672  0.172  0.277    NaN  NaN  NaN  NaN   

   IFFB%  HR/FB  IFH%  BUH%   wOBA  wRAA   wRC  Bat  Fld  Rep  Pos  RAR  WAR  \
0    NaN    NaN   NaN   NaN  0.328  -2.7  29.0 -4.2 -5.0  7.8 -1.6 -2.1 -0.2   
1    NaN    NaN   NaN   NaN  0.289  -7.8  15.0 -7.7 -2.0  5.6  1.5 -2.6 -0.3   
2    NaN    NaN   NaN   NaN  0.289  -7.8  15.0 -7.7 -2.0  5.6  1.5 -2.6 -0.3   

   Dol  Spd  wRC+   WPA  -WPA  +WPA   RE24   REW   pLI  phLI    PH  WPA/LI  \
0  NaN  3.9  87.0  0.37 -3.89  4.26  -6.74 -0.57  0.90  1.79  14.0   -0.16   
1  NaN  4.2  67.0 -0.40 -3.05  2.65 -12.42 -1.20  0.95  1.12  16.0   -0.88   
2  NaN  4.2  67.0 -0.40 -3.05  2.65 -12.42 -1.20  0.95  1.12  16.0   -0.88   

   Clutch  FB% (Pitch)  FBv  SL%  SLv  CT%  CTv  CB%  CBv  CH%  CHv  SF%  SFv  \
0    0.57          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1    0.46          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
2    0.46          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   KN%  KNv  XX%  PO%  wFB  wSL  wCT  wCB  wCH  wSF  wKN  wFB/C  wSL/C  wCT/C  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN    NaN    NaN    NaN   
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN    NaN    NaN    NaN   
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN    NaN    NaN    NaN   

   wCB/C  wCH/C  wSF/C  wKN/C  O-Swing%  Z-Swing%  Swing%  O-Contact%  \
0    NaN    NaN    NaN    NaN       NaN       NaN     NaN         NaN   
1    NaN    NaN    NaN    NaN       NaN       NaN     NaN         NaN   
2    NaN    NaN    NaN    NaN       NaN       NaN     NaN         NaN   

   Z-Contact%  Contact%  Zone%  F-Strike%  SwStr%  BsR  FA% (sc)  FT% (sc)  \
0         NaN       NaN    NaN        NaN     NaN  0.0       NaN       NaN   
1         NaN       NaN    NaN        NaN     NaN -0.2       NaN       NaN   
2         NaN       NaN    NaN        NaN     NaN -0.2       NaN       NaN   

   FC% (sc)  FS% (sc)  FO% (sc)  SI% (sc)  SL% (sc)  CU% (sc)  KC% (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   EP% (sc)  CH% (sc)  SC% (sc)  KN% (sc)  UN% (sc)  vFA (sc)  vFT (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2       NaN       NaN       NaN       NaN       NaN       NaN       NaN   

   vFC (sc)  vFS (sc)  vFO (sc)  vSI (sc)  vSL (sc)  vCU (sc)  vKC (sc)  \
0       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
1       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
2       NaN       NaN     

In [64]:
# check missing values
batting_adv.isna().sum()

Season      381
Name        381
POS         134
Team          0
Age         381
          ...  
CStr%      2098
CSW%       2098
xBA       18737
xSLG      18737
xwOBA     18737
Length: 320, dtype: int64

**Handling missing values in `batting_adv`**

Unfortunately, we have many missing values in the `batting_adv` table. This is expected, since there is a lot of variance when it comes to collecting advanced StatCast data. 

Lets limit `batting_adv` to data between 2014 and 2021 and fill in the remaining missing values with 0s. 

In [65]:
batting_adv = batting_adv[batting_adv['Season'] >= 2014]

In [66]:
# remove rows where stats were not scraped properly (same as in batting_basic table)
# rows where 
batting_adv = batting_adv[batting_adv['Age'].notna()]

In [67]:
# fill all missing NaNs with 0
batting_adv = batting_adv.fillna(0)

In [68]:
# remove repeating total statistics for one season for traded players, keep first entry
batting_adv = batting_adv.drop_duplicates(['Name', 'Season'], keep = 'first')

In [69]:
batting_adv.shape

(3292, 320)

In [70]:
batting_adv.head()

Season              Name POS Team   Age     Salary      G     AB     PA  \
25  2014.0       Bobby Abreu  OF  NYM  40.0  9000000.0   78.0  133.0  155.0   
51  2014.0     Dustin Ackley  OF  SEA  26.0  1700000.0  143.0  502.0  542.0   
52  2015.0     Dustin Ackley  2B  SEA  27.0  2600000.0  108.0  238.0  264.0   
57  2018.0  Ronald Acuna Jr.  OF  ATL  20.0   545000.0  111.0  433.0  487.0   
58  2019.0  Ronald Acuna Jr.  OF  ATL  21.0   560000.0  156.0  626.0  715.0   

        H     1B    2B   3B    HR      R    RBI    BB  IBB     SO  HBP   SF  \
25   33.0   23.0   9.0  0.0   1.0   12.0   14.0  20.0  0.0   21.0  0.0  2.0   
51  123.0   78.0  27.0  4.0  14.0   64.0   65.0  32.0  1.0   90.0  3.0  2.0   
52   55.0   31.0  11.0  3.0  10.0   28.0   30.0  18.0  0.0   45.0  1.0  4.0   
57  127.0   71.0  26.0  4.0  26.0   78.0   64.0  45.0  2.0  123.0  6.0  3.0   
58  175.0  110.0  22.0  2.0  41.0  127.0  101.0  76.0  4.0  188.0  9.0  1.0   

     SH   GDP    SB   CS    AVG     GB     FB     LD  IFFB  Pitches   Balls  \
25  0.0   3.0   1.0  0.0  0.248   56.0   33.0   25.0   0.0    647.0   260.0   
51  3.0  10.0   8.0  4.0  0.245  187.0  150.0   76.0  14.0   2054.0   730.0   
52  3.0   3.0   2.0  2.0  0.231   86.0   79.0   31.0   7.0   1073.0   415.0   
57  0.0   4.0  16.0  5.0  0.293  132.0  123.0   57.0  11.0   1985.0   782.0   
58  0.0   8.0  37.0  9.0  0.280  167.0  165.0  107.0   8.0   3048.0  1204.0   

    Strikes   IFH   BU  BUH    BB%     K%  BB/K    OBP    SLG    OPS    ISO  \
25    387.0   0.0  0.0  0.0  0.129  0.135  0.95  0.342  0.338  0.680  0.090   
51   1324.0   7.0  4.0  0.0  0.059  0.166  0.36  0.293  0.398  0.692  0.153   
52    658.0   2.0  4.0  1.0  0.068  0.170  0.40  0.284  0.429  0.712  0.197   
57   1203.0  11.0  1.0  0.0  0.092  0.253  0.37  0.366  0.552  0.917  0.259   
58   1844.0  19.0  0.0  0.0  0.106  0.263  0.40  0.365  0.518  0.883  0.238   

    BABIP   GB/FB    LD%    GB%    FB%  IFFB%  HR/FB   IFH%  BUH%   wOBA  \
25  0.283  0.0170  0.219  0.491  0.289  0.000  0.030  0.000  0.00  0.309   
51  0.273  0.0125  0.184  0.453  0.363  0.093  0.093  0.037  0.00  0.305   
52  0.241  0.0109  0.158  0.439  0.403  0.089  0.127  0.023  0.25  0.305   
57  0.352  0.0107  0.183  0.423  0.394  0.089  0.211  0.083  0.00  0.388   
58  0.337  0.0101  0.244  0.380  0.376  0.048  0.248  0.114  0.00  0.369   

    wRAA    wRC   Bat  Fld   Rep  Pos   RAR  WAR     Dol  Spd   wRC+   WPA  \
25  -0.1   17.0   0.0 -4.7   4.4 -1.4  -2.0 -0.2  ($1.6)  2.4  100.0  0.23   
51  -2.0   56.0  -1.2  0.8  15.3 -6.7  11.4  1.3    $9.5  4.9   98.0  0.06   
52  -1.8   28.0  -1.8 -3.0   7.7 -2.3  -2.2 -0.2  ($1.9)  4.5   94.0 -0.92   
57  28.9   86.0  25.2  0.7  14.6 -4.4  39.4  4.1   $32.4  6.3  142.0  1.96   
58  30.3  120.0  24.0  3.7  22.5 -1.3  57.3  5.6   $44.5  6.2  125.0  4.50   

     -WPA   +WPA   RE24   REW   pLI  phLI    PH  WPA/LI  Clutch  FB% (Pitch)  \
25  -3.06   3.28   2.82  0.36  1.26  1.60  43.0    0.26   -0.08        0.637   
51  -9.42   9.47   3.68  0.36  0.96  1.60   6.0   -0.05    0.11        0.611   
52  -5.09   4.17  -6.14 -0.64  1.07  1.32  19.0    0.12   -0.98        0.586   
57  -7.60   9.56  26.69  2.80  0.92  0.44   2.0    2.26   -0.12        0.529   
58 -11.05  15.55  41.83  4.11  0.95  1.20   2.0    3.61    1.13        0.562   

     FBv    SL%   SLv    CT%   CTv    CB%   CBv    CH%   CHv    SF%   SFv  \
25  92.5  0.099  83.8  0.053  88.1  0.079  78.0  0.124  83.7  0.009  87.2   
51  91.8  0.114  82.7  0.043  87.2  0.100  76.9  0.100  83.8  0.021  84.9   
52  92.5  0.089  85.0  0.052  88.8  0.092  77.8  0.125  84.0  0.042  84.3   
57  93.4  0.222  85.5  0.058  89.7  0.095  79.5  0.085  84.1  0.010  86.1   
58  93.8  0.196  85.5  0.061  89.2  0.085  80.3  0.086  84.8  0.010  86.5   

      KN%   KNv    XX%  PO%   wFB  wSL  wCT  wCB  wCH  wSF  wKN  wFB/C  wSL/C  \
25  0.000   0.0  0.000  0.0   2.0 -0.1 -0.1 -0.3 -1.2 -0.1  0.0   0.49  -0.13   
51  0.012  75.7  0.002  0.0  -5.6  0.6  0.0  1.1 -2.4  1.3  0.1  -0.44  

Replace duplicate team names

In [71]:
# team names to replace 
replacement_names = {'ANA':'LAA', 'FLA':'MIA', 'MON':'WSN','TBD':'TBR'}

# replace the above names in Team
batting_adv['Team'] = batting_adv['Team'].replace(replacement_names)

In [72]:
display(batting_adv['Team'].unique())
len(batting_adv['Team'].unique())

array(['NYM', 'SEA', 'ATL', 'COL', 'TBR', 'MIL', 'STL', 'WSN', 'TEX',
       'DET', 'LAA', 'SFG', 'MIN', 'MIA', 'ARI', 'CIN', 'BAL', 'KCR',
       'CHC', 'PHI', 'PIT', 'CLE', 'SDP', 'OAK', 'CHW', 'HOU', 'NYY',
       'BOS', 'LAD', 'TOR'], dtype=object)

30

**Adjusting for Inflation in 2021**

In [73]:
# join to cpi
batting_adv = batting_adv.merge(cpi, left_on='Season', right_on='Year', how='left')

In [74]:
batting_adv.columns

Index(['Season', 'Name', 'POS', 'Team', 'Age', 'Salary', 'G', 'AB', 'PA', 'H',
       ...
       'HardHit', 'HardHit%', 'Events', 'CStr%', 'CSW%', 'xBA', 'xSLG',
       'xwOBA', 'Year', 'CPI'],
      dtype='object', length=322)

In [75]:
# create new adjusted 2021 salary using 2021 CPI to determine specific year salary
batting_adv['Adjusted 2021 Salary'] = round((batting_adv['Salary'] * 262.2) / batting_adv['CPI'],0)

In [77]:
# drop extra columns
batting_adv = batting_adv.drop(columns = ['Year', 'CPI'])

In [79]:
# remove adjusted salary to beginning of df
adjusted_adv_sal = batting_adv['Adjusted 2021 Salary']
batting_adv.drop(labels=['Adjusted 2021 Salary'], axis=1, inplace = True)
batting_adv.insert(6, 'Adjusted 2021 Salary', adjusted_adv_sal)
batting_adv

Season              Name POS Team   Age     Salary  \
0     2014.0       Bobby Abreu  OF  NYM  40.0  9000000.0   
1     2014.0     Dustin Ackley  OF  SEA  26.0  1700000.0   
2     2015.0     Dustin Ackley  2B  SEA  27.0  2600000.0   
3     2018.0  Ronald Acuna Jr.  OF  ATL  20.0   545000.0   
4     2019.0  Ronald Acuna Jr.  OF  ATL  21.0   560000.0   
...      ...               ...  ..  ...   ...        ...   
3287  2016.0       Mike Zunino   C  SEA  25.0   523500.0   
3288  2017.0       Mike Zunino   C  SEA  26.0   570000.0   
3289  2018.0       Mike Zunino   C  SEA  27.0  2975000.0   
3290  2019.0       Mike Zunino   C  TBR  28.0  4412000.0   
3291  2021.0       Mike Zunino   C  TBR  30.0  2000000.0   

      Adjusted 2021 Salary      G     AB     PA      H     1B    2B   3B  \
0               10029411.0   78.0  133.0  155.0   33.0   23.0   9.0  0.0   
1                1894444.0  143.0  502.0  542.0  123.0   78.0  27.0  4.0   
2                2904063.0  108.0  238.0  264.0   55.0   31.0  11.0  3.0   
3                 574485.0  111.0  433.0  487.0  127.0   71.0  26.0  4.0   
4                 581582.0  156.0  626.0  715.0  175.0  110.0  22.0  2.0   
...                    ...    ...    ...    ...    ...    ...   ...  ...   
3287              577574.0   55.0  164.0  192.0   34.0   15.0   7.0  0.0   
3288              613477.0  124.0  387.0  435.0   97.0   47.0  25.0  0.0   
3289             3135948.0  113.0  373.0  405.0   75.0   37.0  18.0  0.0   
3290             4582035.0   90.0  266.0  289.0   44.0   24.0  10.0  1.0   
3291             2000000.0  109.0  333.0  375.0   72.0   26.0  11.0  2.0   

        HR      R    RBI    BB  IBB     SO  HBP   SF   SH   GDP    SB   CS  \
0      1.0   12.0   14.0  20.0  0.0   21.0  0.0  2.0  0.0   3.0   1.0  0.0   
1     14.0   64.0   65.0  32.0  1.0   90.0  3.0  2.0  3.0  10.0   8.0  4.0   
2     10.0   28.0   30.0  18.0  0.0   45.0  1.0  4.0  3.0   3.0   2.0  2.0   
3     26.0   78.0   64.0  45.0  2.0  123.0  6.0  3.0  0.0   4.0  16.0  5.0   
4     41.0  127.0  101.0  76.0  4.0  188.0  9.0  1.0  0.0   8.0  37.0  9.0   
...    ...    ...    ...   ...  ...    ...  ...  ...  ...   ...   ...  ...   
3287  12.0   16.0   31.0  21.0  0.0   65.0  6.0  1.0  0.0   0.0   0.0  0.0   
3288  25.0   52.0   64.0  39.0  0.0  160.0  8.0  1.0  0.0   8.0   1.0  0.0   
3289  20.0   37.0   44.0  24.0  0.0  150.0  6.0  2.0  0.0   7.0   0.0  0.0   
3290   9.0   30.0   32.0  20.0  0.0   98.0  3.0  0.0  0.0   4.0   0.0  0.0   
3291  33.0   64.0   62.0  34.0  0.0  132.0  7.0  1.0  0.0   7.0   0.0  0.0   

        AVG     GB     FB     LD  IFFB  Pitches   Balls  Strikes   IFH   BU  \
0     0.248   56.0   33.0   25.0   0.0    647.0   260.0    387.0   0.0  0.0   
1     0.245  187.0  150.0   76.0  14.0   2054.0   730.0   1324.0   7.0  4.0   
2     0.231   86.0   79.0   31.0   7.0   1073.0   415.0    658.0   2.0  4.0   
3     0.293  132.0  123.0   57.0  11.0   1985.0   782.0   1203.0  11.0  1.0   
4     0.280  167.0  165.0  107.0   8.0   3048.0  1204.0   1844.0  19.0  0.0   
...     ...    ...    ...    ...   ...      ...     ...      ...   ...  ...   
3287  0.207   29.0   52.0   18.0   6.0    810.0   313.0    497.0   3.0  1.0   
3288  0.251   73.0  104.0   51.0  13.0   1800.0   662.0   1138.0   3.0  0.0   
3289  0.201   84.0   99.0   42.0  13.0   1714.0   625.0   1089.0   6.0  0.0   
3290  0.165   68.0   77.0   23.0  14.0   1189.0   415.0    774.0   3.0  0.0   
3291  0.216   61.0  109.0   32.0  25.0   1502.0   566.0    936.0   5.0  0.0   

      BUH    BB%     K%  BB/K    OBP    SLG    OPS    ISO  BABIP   GB/FB  \
0     0.0  0.129  0.135  0.95  0.342  0.338  0.680  0.090  0.283  0.0170   
1     0.0  0.059  0.166  0.36  0.293  0.398  0.692  0.153  0.273  0.0125   
2     1.0  0.068  0.170  0.40  0.284  0.429  0.712  0.197  0.241  0.0109   
3     0.0  0.092  0.253  0.37  0.366  0.552  0.917  0.259  0.352  0.0107   
4     0.0  0.106  0.263  0.40  0.365  0.518  0.883  0.238  0.337  0.0101   
...   ...    ...    ...   ... 

**Save `batting_adv` as a csv**

In [80]:
path = "Data/"

batting_adv.to_csv(path + 'batting_adv')

## Formatting the Pitching Table

**Join `pitching_2021` with `bwar_pitch_2021` on the following columns:**

**`pitching_2021`** 
- `Name`
- `Season`

**`bwar_pitch_2021`**  
- `name_common`
- `year_ID`

In [ ]:
# join pitching_2021 onto bwar_pitch_2021 (left join)
pitching_merged = pd.merge(bwar_pitch_2021, pitching_2021, how='left',left_on=['name_common','year_ID'],right_on=['Name','Season'])

In [ ]:
# check
pitching_merged

**Filling missing `salary` values.**

Similar to filling in missing `salary` values for the batting table, we will apply the same function to the `pitching_merged` table. 

To fill in these missing values, we will assume the following:
- A player's salary is the same across consecutive seasons even if the salary is missing
- Players traded mid-season will have the same salary for that season across both teams


In [ ]:
# make a copy and specify which column to fill missing values
salary_col = ['salary']
pitching_filled = pitching_merged.copy()

In [ ]:
# forward and backward fills missing values grouped by name
pitching_filled.loc[:,salary_col] = pitching_merged.loc[:,salary_col].groupby(pitching_filled['name_common']).ffill().bfill()

In [ ]:
# check
pitching_filled['salary'].isna().sum()

In [ ]:
# check
pitching_merged[['name_common', 'year_ID', 'salary']].head(15)

In [ ]:
# check
pitching_filled[['name_common', 'year_ID', 'salary']].head(15)

In [ ]:
pitching_filled

## Simplify and separate the `pitching_filled` table

Similar to what we did above to the `batting_filled` table, we will now simplify into two separate tables for `pitching_filled`. 

Thus, we will create two separate pitching tables:
- `pitching_basic` (containing basic data between 2000 - 2021)
- `pitching_adv` (containing basic AND StatCast data between 2014 - 2021)

We'll also remove repetitive columns that we brought in during the initial merge of tables. 

In [ ]:
pitching_filled.head(3)

In [ ]:
# get all column names
with pd.option_context('display.max_seq_items', None):
    print (pitching_filled.columns)

**Specify columns for `pitching_basic`**

In [ ]:
# basic pitching columns to keep
pitching_basic_columns = ['Season', 'Name', 'team_ID', 'Age', 'salary', 'W', 'L',
       'WAR', 'ERA', 'G', 'GS', 'CG', 'ShO', 'SV', 'BS', 'IP', 'TBF', 'H', 'R',
       'ER', 'HR', 'BB', 'IBB', 'HBP', 'WP', 'BK', 'SO', 'GB', 'FB', 'LD',
       'IFFB', 'Balls', 'Strikes', 'Pitches']

# create new pitching_basic table with specified columns i want
pitching_basic = pitching_filled[pitching_basic_columns]

# renaming columns
pitching_basic = pitching_basic.rename(columns = {'salary': 'Salary',
                                                  'team_ID': 'Team'})

**Specify columns for `pitching_adv`**

In [ ]:
# adv pitching columns to keep
pitching_adv_columns = ['Season', 'Name', 'team_ID', 'Age', 'salary', 'W', 'L',
       'WAR', 'ERA', 'G', 'GS', 'CG', 'ShO', 'SV', 'BS', 'IP', 'TBF', 'H', 'R',
       'ER', 'HR', 'BB', 'IBB', 'HBP', 'WP', 'BK', 'SO', 'GB', 'FB', 'LD',
       'IFFB', 'Balls', 'Strikes', 'Pitches', 'RS', 'IFH', 'BU', 'BUH', 'K/9',
       'BB/9', 'K/BB', 'H/9', 'HR/9', 'AVG', 'WHIP', 'BABIP', 'LOB%', 'FIP',
       'GB/FB', 'LD%', 'GB%', 'FB%', 'IFFB%', 'HR/FB', 'IFH%', 'BUH%',
       'Starting', 'Start-IP', 'Relieving', 'Relief-IP', 'RAR', 'Dollars',
       'tERA', 'xFIP', 'WPA', '-WPA', '+WPA', 'RE24', 'REW', 'pLI', 'inLI',
       'gmLI', 'exLI', 'Pulls', 'WPA/LI', 'Clutch', 'FB% 2', 'FBv', 'SL%',
       'SLv', 'CT%', 'CTv', 'CB%', 'CBv', 'CH%', 'CHv', 'SF%', 'SFv', 'KN%',
       'KNv', 'XX%', 'PO%', 'wFB', 'wSL', 'wCT', 'wCB', 'wCH', 'wSF', 'wKN',
       'wFB/C', 'wSL/C', 'wCT/C', 'wCB/C', 'wCH/C', 'wSF/C', 'wKN/C',
       'O-Swing%', 'Z-Swing%', 'Swing%', 'O-Contact%', 'Z-Contact%',
       'Contact%', 'Zone%', 'F-Strike%', 'SwStr%', 'HLD', 'SD', 'MD', 'ERA-',
       'FIP-', 'xFIP-', 'K%', 'BB%', 'SIERA', 'RS/9', 'E-F', 'FA% (sc)',
       'FT% (sc)', 'FC% (sc)', 'FS% (sc)', 'FO% (sc)', 'SI% (sc)', 'SL% (sc)',
       'CU% (sc)', 'KC% (sc)', 'EP% (sc)', 'CH% (sc)', 'SC% (sc)', 'KN% (sc)',
       'UN% (sc)', 'vFA (sc)', 'vFT (sc)', 'vFC (sc)', 'vFS (sc)', 'vFO (sc)',
       'vSI (sc)', 'vSL (sc)', 'vCU (sc)', 'vKC (sc)', 'vEP (sc)', 'vCH (sc)',
       'vSC (sc)', 'vKN (sc)', 'FA-X (sc)', 'FT-X (sc)', 'FC-X (sc)',
       'FS-X (sc)', 'FO-X (sc)', 'SI-X (sc)', 'SL-X (sc)', 'CU-X (sc)',
       'KC-X (sc)', 'EP-X (sc)', 'CH-X (sc)', 'SC-X (sc)', 'KN-X (sc)',
       'FA-Z (sc)', 'FT-Z (sc)', 'FC-Z (sc)', 'FS-Z (sc)', 'FO-Z (sc)',
       'SI-Z (sc)', 'SL-Z (sc)', 'CU-Z (sc)', 'KC-Z (sc)', 'EP-Z (sc)',
       'CH-Z (sc)', 'SC-Z (sc)', 'KN-Z (sc)', 'wFA (sc)', 'wFT (sc)',
       'wFC (sc)', 'wFS (sc)', 'wFO (sc)', 'wSI (sc)', 'wSL (sc)', 'wCU (sc)',
       'wKC (sc)', 'wEP (sc)', 'wCH (sc)', 'wSC (sc)', 'wKN (sc)',
       'wFA/C (sc)', 'wFT/C (sc)', 'wFC/C (sc)', 'wFS/C (sc)', 'wFO/C (sc)',
       'wSI/C (sc)', 'wSL/C (sc)', 'wCU/C (sc)', 'wKC/C (sc)', 'wEP/C (sc)',
       'wCH/C (sc)', 'wSC/C (sc)', 'wKN/C (sc)', 'O-Swing% (sc)',
       'Z-Swing% (sc)', 'Swing% (sc)', 'O-Contact% (sc)', 'Z-Contact% (sc)',
       'Contact% (sc)', 'Zone% (sc)', 'Pace', 'RA9-WAR', 'BIP-Wins',
       'LOB-Wins', 'FDP-Wins', 'Age Rng', 'K-BB%', 'Pull%', 'Cent%', 'Oppo%',
       'Soft%', 'Med%', 'Hard%', 'kwERA', 'TTO%', 'CH% (pi)', 'CS% (pi)',
       'CU% (pi)', 'FA% (pi)', 'FC% (pi)', 'FS% (pi)', 'KN% (pi)', 'SB% (pi)',
       'SI% (pi)', 'SL% (pi)', 'XX% (pi)', 'vCH (pi)', 'vCS (pi)', 'vCU (pi)',
       'vFA (pi)', 'vFC (pi)', 'vFS (pi)', 'vKN (pi)', 'vSB (pi)', 'vSI (pi)',
       'vSL (pi)', 'vXX (pi)', 'CH-X (pi)', 'CS-X (pi)', 'CU-X (pi)',
       'FA-X (pi)', 'FC-X (pi)', 'FS-X (pi)', 'KN-X (pi)', 'SB-X (pi)',
       'SI-X (pi)', 'SL-X (pi)', 'XX-X (pi)', 'CH-Z (pi)', 'CS-Z (pi)',
       'CU-Z (pi)', 'FA-Z (pi)', 'FC-Z (pi)', 'FS-Z (pi)', 'KN-Z (pi)',
       'SB-Z (pi)', 'SI-Z (pi)', 'SL-Z (pi)', 'XX-Z (pi)', 'wCH (pi)',
       'wCS (pi)', 'wCU (pi)', 'wFA (pi)', 'wFC (pi)', 'wFS (pi)', 'wKN (pi)',
       'wSB (pi)', 'wSI (pi)', 'wSL (pi)', 'wXX (pi)', 'wCH/C (pi)',
       'wCS/C (pi)', 'wCU/C (pi)', 'wFA/C (pi)', 'wFC/C (pi)', 'wFS/C (pi)',
       'wKN/C (pi)', 'wSB/C (pi)', 'wSI/C (pi)', 'wSL/C (pi)', 'wXX/C (pi)',
       'O-Swing% (pi)', 'Z-Swing% (pi)', 'Swing% (pi)', 'O-Contact% (pi)',
       'Z-Contact% (pi)', 'Contact% (pi)', 'Zone% (pi)', 'Pace (pi)', 'FRM',
       'K/9+', 'BB/9+', 'K/BB+', 'H/9+', 'HR/9+', 'AVG+', 'WHIP+', 'BABIP+',
       'LOB%+', 'K%+', 'BB%+', 'LD%+', 'GB%+', 'FB%+', 'HR/FB%+', 'Pull%+',
       'Cent%+', 'Oppo%+', 'Soft%+', 'Med%+', 'Hard%+', 'EV', 'LA', 'Barrels',
       'Barrel%', 'maxEV', 'HardHit', 'HardHit%', 'Events', 'CStr%', 'CSW%',
       'xERA']

# create new batting_adv table with specified columns i want
pitching_adv = pitching_filled[pitching_adv_columns]

# renaming columns
pitching_adv = pitching_adv.rename(columns = {'salary': 'Salary',
                                              'team_ID': 'Team'})

**Checking and cleaning `pitching_basic` first:**

In [ ]:
# check
pitching_basic.head()

In [ ]:
pitching_basic.info()

In [ ]:
# check missing values
pitching_basic.isna().sum()

**Handling missing values in `pitching_basic`**

There are 5065 missing values for standard pitching statistics likely attributed to inconsistencies when merging from earlier. 

We'll just drop these missing rows from `pitching_basic`. 

In [ ]:
# drop all missing nan rows
pitching_basic = pitching_basic.dropna()

In [ ]:
# check all are 0 (note: columns get cut off as shown below)
pd.set_option('display.max_rows', 10) # change to None show all columns 
pitching_basic.isna().sum()

In [ ]:
# remove repeating total statistics for one season for traded players, keep first entry
pitching_basic = pitching_basic.drop_duplicates(['Name', 'Season'], keep = 'first')

Replace duplicate team names

In [ ]:
# team names to replace 
replacement_names = {'ANA':'LAA', 'FLA':'MIA', 'MON':'WSN','TBD':'TBR'}

# replace the above names in Team
pitching_basic['Team'] = pitching_basic['Team'].replace(replacement_names)

display(pitching_basic['Team'].unique())
len(pitching_basic['Team'].unique())

In [ ]:
# final pitching basic dimensions
pitching_basic.shape

In [ ]:
pitching_basic.head()

**Save `pitching_basic` as a csv**

In [ ]:
path = "Data/"

pitching_basic.to_csv(path + 'pitching_basic')

**Checking and cleaning `pitching_adv`:**

In [ ]:
pitching_adv.head(3)

**Handling missing values in `pitching_adv`**

Unfortunately, we have many missing values in the `pitching_adv` table. This is expected, since there is a lot of variance when it comes to collecting advanced StatCast data. 

Lets limit `pitching_adv` to data between 2014 and 2021 and fill in the remaining missing values with 0s. 

In [ ]:
pitching_adv = pitching_adv[pitching_adv['Season'] >= 2014]

In [ ]:
# remove rows where stats were not scraped properly (same as in batting_basic table)
pitching_adv = pitching_adv[pitching_adv['Age'].notna()]

In [ ]:
# fill all missing NaNs with 0
pitching_adv = pitching_adv.fillna(0)

In [ ]:
# check all are 0 (note: columns get cut off as shown below)
pd.set_option('display.max_rows', 10) # change to None show all columns 
pitching_adv.isna().sum()

In [ ]:
# remove repeating total statistics for one season for traded players, keep first entry
pitching_adv = pitching_adv.drop_duplicates(['Name', 'Season'], keep = 'first')

Replace duplicate team names

In [ ]:
# team names to replace 
replacement_names = {'ANA':'LAA', 'FLA':'MIA', 'MON':'WSN','TBD':'TBR'}

# replace the above names in Team
pitching_adv['Team'] = pitching_adv['Team'].replace(replacement_names)

display(pitching_adv['Team'].unique())
len(pitching_adv['Team'].unique())

In [ ]:
pitching_adv.shape

In [ ]:
pitching_adv.head()

**Save `pitching_adv` as a csv**

In [ ]:
path = "Data/"

pitching_adv.to_csv(path + 'pitching_adv')